# Factored Datathon 2026 — LATAM Bank Dataset Explorer

This notebook is set up to explore the banking dataset hosted on Amazon S3.

**Objectives:**
1. Connect to the read-only S3 bucket.
2. List the available files.
3. Explore the size and structure of the data.
4. Load only small samples to avoid exhausting memory.
5. Review data quality: nulls, duplicates, and columns.
6. Identify tables useful for building an agentic workflow.

> The dataset has approximately 19 million records distributed across 13 tables, so **it's not advisable to download everything at once**.

## 1. Install dependencies

We'll use `boto3` to connect to Amazon S3 and `pandas` to analyze the data.

In [ ]:
!pip -q install boto3 s3fs pyarrow

## 2. Configure AWS S3 access

For security, this notebook **does not store your Secret Access Key**.

When you run the cell:
- copy the **Access Key ID** from the Data Dictionary PDF;
- copy the **Secret Access Key** from the same PDF.

The bucket and region are already configured.

In [ ]:
import os
import getpass
import boto3
import pandas as pd
from io import BytesIO

BUCKET_NAME = "factored-datathon-2026-s3-157725502942-us-east-2-an"
AWS_REGION = "us-east-2"
DATA_PREFIX = "data/"

aws_access_key_id = input("AWS Access Key ID: ").strip()
aws_secret_access_key = getpass.getpass("AWS Secret Access Key: ").strip()

session = boto3.Session(
    aws_access_key_id=aws_access_key_id,
    aws_secret_access_key=aws_secret_access_key,
    region_name=AWS_REGION,
)

s3 = session.client("s3")

print("Configuration created.")

AWS Access Key ID: AKIASJOJMFXPDNU7JH7C
AWS Secret Access Key: ··········
Configuración creada.


## 3. Test the connection

This cell tries to list objects in the bucket. If it works, you should see files or folders under `data/`.

In [ ]:
response = s3.list_objects_v2(
    Bucket=BUCKET_NAME,
    Prefix=DATA_PREFIX,
    MaxKeys=20,
)

objects = response.get("Contents", [])

if not objects:
    print("No objects found under:", DATA_PREFIX)
else:
    print(f"Found at least {len(objects)} objects. First results:\n")
    for obj in objects:
        print(f"{obj['Key']:<90} {obj['Size']/1024/1024:>10.2f} MB")

Se encontraron al menos 20 objetos. Primeros resultados:

data/branches.csv                                                                                0.09 MB
data/call_center_interactions/year=2023/month=06/day=17/call_center_interactions_20230617.csv       0.07 MB
data/call_center_interactions/year=2023/month=06/day=18/call_center_interactions_20230618.csv       0.07 MB
data/call_center_interactions/year=2023/month=06/day=19/call_center_interactions_20230619.csv       0.16 MB
data/call_center_interactions/year=2023/month=06/day=20/call_center_interactions_20230620.csv       0.14 MB
data/call_center_interactions/year=2023/month=06/day=21/call_center_interactions_20230621.csv       0.13 MB
data/call_center_interactions/year=2023/month=06/day=22/call_center_interactions_20230622.csv       0.12 MB
data/call_center_interactions/year=2023/month=06/day=23/call_center_interactions_20230623.csv       0.15 MB
data/call_center_interactions/year=2023/month=06/day=24/call_center_interactions_

## 4. List all dataset files

This function paginates automatically so it isn't limited to just the first 1,000 objects.

In [ ]:
def list_s3_files(bucket, prefix="data/"):
    paginator = s3.get_paginator("list_objects_v2")
    rows = []

    for page in paginator.paginate(Bucket=bucket, Prefix=prefix):
        for obj in page.get("Contents", []):
            rows.append({
                "key": obj["Key"],
                "size_mb": obj["Size"] / 1024 / 1024,
                "last_modified": obj["LastModified"],
            })

    return pd.DataFrame(rows)

files_df = list_s3_files(BUCKET_NAME, DATA_PREFIX)

print("Number of objects:", len(files_df))
display(files_df.head(50))

Cantidad de objetos: 7671


,key,size_mb,last_modified
0,data/branches.csv,0.085382,2026-09-01 02:36:08+00:00
1,data/call_center_interactions/year=2023/month=...,0.074080,2026-09-01 02:36:10+00:00
2,data/call_center_interactions/year=2023/month=...,0.074688,2026-09-01 02:36:10+00:00
3,data/call_center_interactions/year=2023/month=...,0.159051,2026-09-01 02:36:10+00:00
4,data/call_center_interactions/year=2023/month=...,0.140602,2026-09-01 02:36:10+00:00
5,data/call_center_interactions/year=2023/month=...,0.132975,2026-09-01 02:36:08+00:00
6,data/call_center_interactions/year=2023/month=...,0.119330,2026-09-01 02:36:08+00:00
7,data/call_center_interactions/year=2023/month=...,0.152835,2026-09-01 02:36:08+00:00
8,data/call_center_interactions/year=2023/month=...,0.080551,2026-09-01 02:36:08+00:00
9,data/call_center_interactions/year=2023/month=...,0.057339,2026-09-01 02:36:08+00:00


## 5. Summary by extension and size

This helps understand whether the data comes as CSV, Parquet, or another format.

In [ ]:
if not files_df.empty:
    files_df["extension"] = files_df["key"].str.extract(r"(\.[A-Za-z0-9]+)$", expand=False).fillna("(no extension)")
    display(
        files_df.groupby("extension")
        .agg(files=("key", "count"), total_mb=("size_mb", "sum"))
        .sort_values("total_mb", ascending=False)
    )

,files,total_mb
extension,,
.csv,7671,5101.51146


## 6. Search files by table name

The documented tables include, among others:

- `customers`
- `products`
- `transactions`
- `call_center_interactions`
- `call_transcripts`
- `satisfaction_surveys`
- `digital_events`
- `complaints`
- `campaign_sends`

You can search for matches without downloading anything.

In [ ]:
TABLE_NAMES = [
    "customers",
    "products",
    "transactions",
    "call_center_interactions",
    "call_transcripts",
    "satisfaction_surveys",
    "digital_events",
    "complaints",
    "campaign_sends",
    "daily_exchange_rates",
]

for table in TABLE_NAMES:
    matches = files_df[files_df["key"].str.contains(table, case=False, na=False)]
    print(f"\n{table}: {len(matches)} objects")
    if len(matches):
        display(matches.head(10))


customers: 1 objetos


,key,size_mb,last_modified,extension
4375,data/customers.csv,44.724806,2026-09-01 02:38:54+00:00,.csv



products: 1 objetos


,key,size_mb,last_modified,extension
5475,data/products.csv,65.053602,2026-09-01 02:48:51+00:00,.csv



transactions: 1097 objetos


,key,size_mb,last_modified,extension
6574,data/transactions/year=2023/month=06/day=17/tr...,0.501984,2026-09-01 02:49:26+00:00,.csv
6575,data/transactions/year=2023/month=06/day=18/tr...,0.505768,2026-09-01 02:49:26+00:00,.csv
6576,data/transactions/year=2023/month=06/day=19/tr...,0.721908,2026-09-01 02:49:26+00:00,.csv
6577,data/transactions/year=2023/month=06/day=20/tr...,0.743258,2026-09-01 02:49:26+00:00,.csv
6578,data/transactions/year=2023/month=06/day=21/tr...,0.854450,2026-09-01 02:49:26+00:00,.csv
6579,data/transactions/year=2023/month=06/day=22/tr...,0.778510,2026-09-01 02:49:26+00:00,.csv
6580,data/transactions/year=2023/month=06/day=23/tr...,0.933638,2026-09-01 02:49:26+00:00,.csv
6581,data/transactions/year=2023/month=06/day=24/tr...,0.498577,2026-09-01 02:49:26+00:00,.csv
6582,data/transactions/year=2023/month=06/day=25/tr...,0.496027,2026-09-01 02:49:27+00:00,.csv
6583,data/transactions/year=2023/month=06/day=26/tr...,0.869211,2026-09-01 02:49:27+00:00,.csv



call_center_interactions: 1097 objetos


,key,size_mb,last_modified,extension
1,data/call_center_interactions/year=2023/month=...,0.074080,2026-09-01 02:36:10+00:00,.csv
2,data/call_center_interactions/year=2023/month=...,0.074688,2026-09-01 02:36:10+00:00,.csv
3,data/call_center_interactions/year=2023/month=...,0.159051,2026-09-01 02:36:10+00:00,.csv
4,data/call_center_interactions/year=2023/month=...,0.140602,2026-09-01 02:36:10+00:00,.csv
5,data/call_center_interactions/year=2023/month=...,0.132975,2026-09-01 02:36:08+00:00,.csv
6,data/call_center_interactions/year=2023/month=...,0.119330,2026-09-01 02:36:08+00:00,.csv
7,data/call_center_interactions/year=2023/month=...,0.152835,2026-09-01 02:36:08+00:00,.csv
8,data/call_center_interactions/year=2023/month=...,0.080551,2026-09-01 02:36:08+00:00,.csv
9,data/call_center_interactions/year=2023/month=...,0.057339,2026-09-01 02:36:08+00:00,.csv
10,data/call_center_interactions/year=2023/month=...,0.160464,2026-09-01 02:36:08+00:00,.csv



call_transcripts: 1097 objetos


,key,size_mb,last_modified,extension
1098,data/call_transcripts/year=2023/month=06/day=1...,0.067779,2026-09-01 02:37:09+00:00,.csv
1099,data/call_transcripts/year=2023/month=06/day=1...,0.087402,2026-09-01 02:37:09+00:00,.csv
1100,data/call_transcripts/year=2023/month=06/day=1...,0.147844,2026-09-01 02:37:09+00:00,.csv
1101,data/call_transcripts/year=2023/month=06/day=2...,0.153019,2026-09-01 02:37:09+00:00,.csv
1102,data/call_transcripts/year=2023/month=06/day=2...,0.124538,2026-09-01 02:37:09+00:00,.csv
1103,data/call_transcripts/year=2023/month=06/day=2...,0.114937,2026-09-01 02:37:09+00:00,.csv
1104,data/call_transcripts/year=2023/month=06/day=2...,0.146931,2026-09-01 02:37:09+00:00,.csv
1105,data/call_transcripts/year=2023/month=06/day=2...,0.063638,2026-09-01 02:37:09+00:00,.csv
1106,data/call_transcripts/year=2023/month=06/day=2...,0.059257,2026-09-01 02:37:09+00:00,.csv
1107,data/call_transcripts/year=2023/month=06/day=2...,0.164206,2026-09-01 02:37:09+00:00,.csv



satisfaction_surveys: 1097 objetos


,key,size_mb,last_modified,extension
5476,data/satisfaction_surveys/year=2023/month=06/d...,0.025392,2026-09-01 02:48:51+00:00,.csv
5477,data/satisfaction_surveys/year=2023/month=06/d...,0.025052,2026-09-01 02:48:51+00:00,.csv
5478,data/satisfaction_surveys/year=2023/month=06/d...,0.051445,2026-09-01 02:48:51+00:00,.csv
5479,data/satisfaction_surveys/year=2023/month=06/d...,0.046927,2026-09-01 02:48:52+00:00,.csv
5480,data/satisfaction_surveys/year=2023/month=06/d...,0.043792,2026-09-01 02:48:52+00:00,.csv
5481,data/satisfaction_surveys/year=2023/month=06/d...,0.040174,2026-09-01 02:48:52+00:00,.csv
5482,data/satisfaction_surveys/year=2023/month=06/d...,0.050670,2026-09-01 02:48:52+00:00,.csv
5483,data/satisfaction_surveys/year=2023/month=06/d...,0.027505,2026-09-01 02:48:52+00:00,.csv
5484,data/satisfaction_surveys/year=2023/month=06/d...,0.019324,2026-09-01 02:48:52+00:00,.csv
5485,data/satisfaction_surveys/year=2023/month=06/d...,0.053262,2026-09-01 02:48:52+00:00,.csv



digital_events: 1097 objetos


,key,size_mb,last_modified,extension
4377,data/digital_events/year=2023/month=06/day=17/...,4.341200,2026-09-01 02:38:54+00:00,.csv
4378,data/digital_events/year=2023/month=06/day=18/...,1.410533,2026-09-01 02:38:54+00:00,.csv
4379,data/digital_events/year=2023/month=06/day=19/...,2.453911,2026-09-01 02:38:55+00:00,.csv
4380,data/digital_events/year=2023/month=06/day=20/...,2.960979,2026-09-01 02:38:55+00:00,.csv
4381,data/digital_events/year=2023/month=06/day=21/...,4.031322,2026-09-01 02:38:58+00:00,.csv
4382,data/digital_events/year=2023/month=06/day=22/...,4.125672,2026-09-01 02:39:00+00:00,.csv
4383,data/digital_events/year=2023/month=06/day=23/...,3.020381,2026-09-01 02:39:00+00:00,.csv
4384,data/digital_events/year=2023/month=06/day=24/...,2.146443,2026-09-01 02:39:01+00:00,.csv
4385,data/digital_events/year=2023/month=06/day=25/...,2.431797,2026-09-01 02:39:03+00:00,.csv
4386,data/digital_events/year=2023/month=06/day=26/...,2.836190,2026-09-01 02:39:03+00:00,.csv



complaints: 1097 objetos


,key,size_mb,last_modified,extension
3278,data/complaints/year=2023/month=06/day=17/comp...,0.010068,2026-09-01 02:38:33+00:00,.csv
3279,data/complaints/year=2023/month=06/day=18/comp...,0.008785,2026-09-01 02:38:33+00:00,.csv
3280,data/complaints/year=2023/month=06/day=19/comp...,0.016797,2026-09-01 02:38:33+00:00,.csv
3281,data/complaints/year=2023/month=06/day=20/comp...,0.017870,2026-09-01 02:38:33+00:00,.csv
3282,data/complaints/year=2023/month=06/day=21/comp...,0.020050,2026-09-01 02:38:33+00:00,.csv
3283,data/complaints/year=2023/month=06/day=22/comp...,0.015250,2026-09-01 02:38:33+00:00,.csv
3284,data/complaints/year=2023/month=06/day=23/comp...,0.015225,2026-09-01 02:38:33+00:00,.csv
3285,data/complaints/year=2023/month=06/day=24/comp...,0.007611,2026-09-01 02:38:33+00:00,.csv
3286,data/complaints/year=2023/month=06/day=25/comp...,0.009392,2026-09-01 02:38:33+00:00,.csv
3287,data/complaints/year=2023/month=06/day=26/comp...,0.015573,2026-09-01 02:38:33+00:00,.csv



campaign_sends: 1083 objetos


,key,size_mb,last_modified,extension
2195,data/campaign_sends/year=2023/month=07/day=01/...,0.232161,2026-09-01 02:37:37+00:00,.csv
2196,data/campaign_sends/year=2023/month=07/day=02/...,0.222411,2026-09-01 02:37:37+00:00,.csv
2197,data/campaign_sends/year=2023/month=07/day=03/...,0.305234,2026-09-01 02:37:37+00:00,.csv
2198,data/campaign_sends/year=2023/month=07/day=04/...,0.406342,2026-09-01 02:37:37+00:00,.csv
2199,data/campaign_sends/year=2023/month=07/day=05/...,0.298929,2026-09-01 02:37:37+00:00,.csv
2200,data/campaign_sends/year=2023/month=07/day=06/...,0.278778,2026-09-01 02:37:37+00:00,.csv
2201,data/campaign_sends/year=2023/month=07/day=07/...,0.297106,2026-09-01 02:37:37+00:00,.csv
2202,data/campaign_sends/year=2023/month=07/day=08/...,0.232841,2026-09-01 02:37:37+00:00,.csv
2203,data/campaign_sends/year=2023/month=07/day=09/...,0.196675,2026-09-01 02:37:37+00:00,.csv
2204,data/campaign_sends/year=2023/month=07/day=10/...,0.305003,2026-09-01 02:37:37+00:00,.csv



daily_exchange_rates: 1 objetos


,key,size_mb,last_modified,extension
4376,data/daily_exchange_rates.csv,0.74283,2026-09-01 02:38:54+00:00,.csv


## 7. Load a sample from a file

The function below:
- reads CSV using only the first rows (`nrows`);
- reads the full Parquet file if it's small, or tries to load it directly if reasonable;
- avoids manually downloading the entire dataset.

> For very large, partitioned Parquet files, it's better to query specific partitions later on.

In [ ]:
def load_sample_from_s3(key, nrows=5000):
    print("key!! ", key)
    key_lower = key.lower()

    if key_lower.endswith(".csv"):
        obj = s3.get_object(Bucket=BUCKET_NAME, Key=key)
        return pd.read_csv(obj["Body"], nrows=nrows)

    elif key_lower.endswith(".parquet"):
        # For a first analysis. For huge files, choose a small partition.
        obj = s3.get_object(Bucket=BUCKET_NAME, Key=key)
        data = obj["Body"].read()
        return pd.read_parquet(BytesIO(data))

    else:
        raise ValueError(f"Format not automatically supported: {key}")

# Example:
# sample_key = files_df.iloc[0]["key"]
# sample_df = load_sample_from_s3(sample_key, nrows=5000)
# display(sample_df.head())

## 8. Choose a table and analyze a sample

Start with one of these three:
- `transactions`
- `call_center_interactions`
- `complaints`

They're especially useful for designing automated banking workflows with agents.

In [ ]:
TARGET_TABLE = "transactions"  # change this value if you want

matches = files_df[
    files_df["key"].str.contains(TARGET_TABLE, case=False, na=False)
    & files_df["key"].str.contains(r"\.csv$|\.parquet$", case=False, regex=True, na=False)
].copy()

display(matches.head(20))

if matches.empty:
    print("No compatible files found for:", TARGET_TABLE)
else:
    sample_key = matches.sort_values("size_mb").iloc[0]["key"]
    print("Selected file:", sample_key)

    sample_df = load_sample_from_s3(sample_key, nrows=5000)

    print("\nSample shape:", sample_df.shape)
    display(sample_df.head())

,key,size_mb,last_modified,extension
6574,data/transactions/year=2023/month=06/day=17/tr...,0.501984,2026-09-01 02:49:26+00:00,.csv
6575,data/transactions/year=2023/month=06/day=18/tr...,0.505768,2026-09-01 02:49:26+00:00,.csv
6576,data/transactions/year=2023/month=06/day=19/tr...,0.721908,2026-09-01 02:49:26+00:00,.csv
6577,data/transactions/year=2023/month=06/day=20/tr...,0.743258,2026-09-01 02:49:26+00:00,.csv
6578,data/transactions/year=2023/month=06/day=21/tr...,0.854450,2026-09-01 02:49:26+00:00,.csv
6579,data/transactions/year=2023/month=06/day=22/tr...,0.778510,2026-09-01 02:49:26+00:00,.csv
6580,data/transactions/year=2023/month=06/day=23/tr...,0.933638,2026-09-01 02:49:26+00:00,.csv
6581,data/transactions/year=2023/month=06/day=24/tr...,0.498577,2026-09-01 02:49:26+00:00,.csv
6582,data/transactions/year=2023/month=06/day=25/tr...,0.496027,2026-09-01 02:49:27+00:00,.csv
6583,data/transactions/year=2023/month=06/day=26/tr...,0.869211,2026-09-01 02:49:27+00:00,.csv


Archivo seleccionado: data/transactions/year=2024/month=07/day=28/transactions_20240728.csv
key!!  data/transactions/year=2024/month=07/day=28/transactions_20240728.csv

Shape de la muestra: (2189, 22)


,transaction_id,transaction_date,process_date,product_id,customer_id,transaction_type,transaction_category,amount,currency,amount_usd,...,merchant_name,merchant_category,transaction_country,transaction_city,transaction_status,response_code,is_fraud,fraud_score,latitude,longitude
0,TRX-V05B5G9YM8X9EG8UCIR3,2024-07-28 11:47:27,2024-07-28,PRD-GBM54FA86ZLY,CLI-FOB62JT35LMP,Adjustment,NaN,35957.61,ARS,102.74,...,NaN,NaN,Argentina,Mendoza,Approved,0.0,False,NaN,NaN,NaN
1,TRX-2BX60XQH0C3Q8VJXXIVB,2024-07-28 17:10:01,2024-07-28,PRD-F7H2X936579K,CLI-I5ED9E9X76TW,Purchase,Services,18.10,USD,NaN,...,Servicios Públicos,Services,México,Ciudad de México,Approved,NaN,False,4.26,-0.460053,0.290304
2,TRX-EXHTM0FCIY7NFO1EA6A3,2024-07-28 20:08:34,2024-07-28,PRD-3ZT9EAL6RWZK,CLI-YSD26EH3KP05,Transfer,NaN,6477223.99,COP,1619.31,...,NaN,NaN,Colombia,Cartagena,Approved,0.0,False,4.08,5.117426,-74.759883
3,TRX-WGLQRLQDIF0EIZQ4D5EB,2024-07-28 21:38:21,2024-07-28,PRD-NSQGE2MAMT4G,CLI-QZC64289N9EI,Deposit,NaN,599933.01,ARS,1714.09,...,NaN,NaN,Argentina,Córdoba,Approved,0.0,False,NaN,NaN,NaN
4,TRX-2XG5JUPGD56D3F4V3PYC,2024-07-29 01:48:09,2024-07-28,PRD-AXJQ3V10Q3HS,CLI-38G6ES3ME4GC,Deposit,NaN,1422000.47,ARS,4062.86,...,NaN,NaN,Argentina,Córdoba,Approved,0.0,False,17.88,NaN,NaN


## 9. Quick profile of the sample

This answers:
- what columns exist?
- what data types do they have?
- how many nulls are there?
- how many duplicates?
- what categorical values appear?

In [ ]:
def quick_profile(df):
    print("Shape:", df.shape)
    print("\nColumns:")
    print(df.columns.tolist())

    print("\nTypes:")
    display(df.dtypes.to_frame("dtype"))

    print("\nNulls:")
    nulls = df.isna().sum().sort_values(ascending=False)
    display(nulls[nulls > 0].to_frame("null_count"))

    print("\nDuplicates in the sample:", df.duplicated().sum())

    print("\nStatistics:")
    display(df.describe(include="all").T)

    print("\nFrequent values in categorical columns:")
    for col in df.select_dtypes(include=["object", "category", "bool"]).columns:
        print(f"\n--- {col} ---")
        display(df[col].value_counts(dropna=False).head(15).to_frame("count"))

if "sample_df" in globals():
    quick_profile(sample_df)

Shape: (2189, 22)

Columnas:
['transaction_id', 'transaction_date', 'process_date', 'product_id', 'customer_id', 'transaction_type', 'transaction_category', 'amount', 'currency', 'amount_usd', 'channel', 'branch_id', 'merchant_name', 'merchant_category', 'transaction_country', 'transaction_city', 'transaction_status', 'response_code', 'is_fraud', 'fraud_score', 'latitude', 'longitude']

Tipos:


,dtype
transaction_id,object
transaction_date,object
process_date,object
product_id,object
customer_id,object
transaction_type,object
transaction_category,object
amount,float64
currency,object
amount_usd,float64



Nulos:


,null_count
latitude,1807
longitude,1797
merchant_name,1676
merchant_category,1675
branch_id,1533
transaction_category,1350
amount_usd,1249
fraud_score,409
transaction_city,210
response_code,99



Duplicados en la muestra: 0

Estadísticas:


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
transaction_id,2189,2189,TRX-E2LCKFCILHE5F8Z17E76,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
transaction_date,2189,2165,2024-07-28 11:11:04,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
process_date,2189,1,2024-07-28,2189,NaN,NaN,NaN,NaN,NaN,NaN,NaN
product_id,2189,2183,PRD-YA8Q4WP1FGHQ,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_id,2189,2165,CLI-EGGGEC4I20N7,2,NaN,NaN,NaN,NaN,NaN,NaN,NaN
transaction_type,2189,6,Purchase,536,NaN,NaN,NaN,NaN,NaN,NaN,NaN
transaction_category,839,6,Food,191,NaN,NaN,NaN,NaN,NaN,NaN,NaN
amount,2189.0,NaN,NaN,NaN,2066442.191526,5970053.701044,6.18,418.66,5969.06,877862.88,39662094.99
currency,2189,3,USD,1184,NaN,NaN,NaN,NaN,NaN,NaN,NaN
amount_usd,940.0,NaN,NaN,NaN,1793.770777,2456.991313,6.69,230.2725,467.325,2391.3225,9920.27



Valores frecuentes en columnas categóricas:

--- transaction_id ---


,count
transaction_id,
TRX-E2LCKFCILHE5F8Z17E76,1
TRX-V05B5G9YM8X9EG8UCIR3,1
TRX-2BX60XQH0C3Q8VJXXIVB,1
TRX-EXHTM0FCIY7NFO1EA6A3,1
TRX-WGLQRLQDIF0EIZQ4D5EB,1
TRX-2XG5JUPGD56D3F4V3PYC,1
TRX-HPSSS01BGXD1JXP3EBH1,1
TRX-GFYGTVOTD1UVFZB35D4B,1
TRX-ZLFPOAAZAIA4KJZO1LI4,1



--- transaction_date ---


,count
transaction_date,
2024-07-28 11:11:04,2
2024-07-28 12:05:24,2
2024-07-28 11:12:12,2
2024-07-29 01:35:08,2
2024-07-29 02:35:45,2
2024-07-28 23:25:23,2
2024-07-28 20:29:57,2
2024-07-28 09:27:00,2
2024-07-28 08:40:01,2



--- process_date ---


,count
process_date,
2024-07-28,2189



--- product_id ---


,count
product_id,
PRD-YA8Q4WP1FGHQ,2
PRD-VJHULD2MH8BD,2
PRD-D89LWNYV0UUC,2
PRD-GXANIWAI425C,2
PRD-2NUIU0CNTHWH,2
PRD-1F2RXUGEZ3SI,2
PRD-3WJS4JSSNB8G,1
PRD-6CJUK1AT31E3,1
PRD-KJTECWZ8341L,1



--- customer_id ---


,count
customer_id,
CLI-EGGGEC4I20N7,2
CLI-8G92VUHW8OAC,2
CLI-IRQ4YKU3SHPY,2
CLI-38C73LA8QPH6,2
CLI-Q8MQ2YRPSTO8,2
CLI-7E0QKTVNRHL1,2
CLI-FMMNOBS4G1R6,2
CLI-281M8VTSZWFA,2
CLI-VMPON9YWY4GV,2



--- transaction_type ---


,count
transaction_type,
Purchase,536
Withdrawal,489
Transfer,442
Payment,358
Deposit,291
Adjustment,73



--- transaction_category ---


,count
transaction_category,
NaN,1350
Food,191
Services,177
Other,139
Entertainment,128
Transport,121
Health,83



--- currency ---


,count
currency,
USD,1184
COP,601
ARS,404



--- channel ---


,count
channel,
POS,749
ATM,621
Web,353
App,349
Branch,67
Transfer,50



--- branch_id ---


,count
branch_id,
NaN,1533
SUC-LWQND6S8,7
SUC-1C51KLB3,6
SUC-57LRN6RH,5
SUC-M447M9GU,5
SUC-ADIYTS49,5
SUC-12NKB9OB,5
SUC-4B1UVBW3,5
SUC-JDSJM3BZ,4



--- merchant_name ---


,count
merchant_name,
NaN,1676
Servicios Públicos,36
Mercado Central,35
Tienda Don José,31
Cable TV,28
Super Ahorro,27
Internet Plus,26
Restaurante El Buen Sabor,25
Empresa Telefónica,25



--- merchant_category ---


,count
merchant_category,
NaN,1675
Food,119
Services,117
Other,80
Entertainment,76
Transport,66
Health,56



--- transaction_country ---


,count
transaction_country,
México,998
Colombia,671
Argentina,439
Spain,24
Mexico,23
Brazil,18
USA,16



--- transaction_city ---


,count
transaction_city,
NaN,210
Puebla,160
Tijuana,159
Monterrey,157
Querétaro,153
Ciudad de México,151
Guadalajara,146
Cartagena,141
Cali,127



--- transaction_status ---


,count
transaction_status,
Approved,2020
Declined,108
Pending,40
Reversed,21



--- is_fraud ---


,count
is_fraud,
False,2186
True,3


## 10. Specific analysis of `transactions`

According to the Data Dictionary, this table contains fields such as:
- transaction type and category;
- amount and currency;
- channel;
- merchant;
- country/city;
- status;
- fraud indicator and score.

This can reveal workflows such as:
- unrecognized transaction;
- declined transaction;
- suspected fraud;
- reversal or dispute.

In [ ]:
if "sample_df" in globals() and TARGET_TABLE == "transactions":
    interesting_cols = [
        "transaction_type",
        "transaction_category",
        "channel",
        "transaction_status",
        "is_fraud",
        "transaction_country",
        "merchant_category",
    ]

    for col in interesting_cols:
        if col in sample_df.columns:
            print(f"\n### {col}")
            display(sample_df[col].value_counts(dropna=False).head(20).to_frame("count"))

    if "amount_usd" in sample_df.columns:
        print("\nDistribution of amount_usd:")
        display(sample_df["amount_usd"].describe().to_frame().T)


### transaction_type


,count
transaction_type,
Purchase,536
Withdrawal,489
Transfer,442
Payment,358
Deposit,291
Adjustment,73



### transaction_category


,count
transaction_category,
NaN,1350
Food,191
Services,177
Other,139
Entertainment,128
Transport,121
Health,83



### channel


,count
channel,
POS,749
ATM,621
Web,353
App,349
Branch,67
Transfer,50



### transaction_status


,count
transaction_status,
Approved,2020
Declined,108
Pending,40
Reversed,21



### is_fraud


,count
is_fraud,
False,2186
True,3



### transaction_country


,count
transaction_country,
México,998
Colombia,671
Argentina,439
Spain,24
Mexico,23
Brazil,18
USA,16



### merchant_category


,count
merchant_category,
NaN,1675
Food,119
Services,117
Other,80
Entertainment,76
Transport,66
Health,56



Distribución de amount_usd:


,count,mean,std,min,25%,50%,75%,max
amount_usd,940.0,1793.770777,2456.991313,6.69,230.2725,467.325,2391.3225,9920.27


## 11. Next step: choose a workflow

After running the cells, try to answer:

1. Which problems appear most frequently?
2. Which fields enable a decision to be made?
3. Which steps could an agent execute automatically?
4. At what point should it ask for human confirmation?
5. What additional tables would it need to query?

### Example workflow

**Customer:** "I don't recognize this purchase."

**Possible flow:**

`customer message`
→ detect intent
→ identify customer
→ look up transaction
→ verify amount/merchant/location/fraud_score
→ review previous interactions
→ decide whether confirmation is required
→ create/escalate claim
→ respond to customer

This type of flow leverages several tables and demonstrates an **AI-first / agentic** system far better than a simple classifier.

## 12. Optional cleanup of session credentials

Run this cell when you're done if you want to remove the access variables from the notebook environment.

In [ ]:
aws_access_key_id = None
aws_secret_access_key = None
session = None
s3 = None
print("Credentials removed from the notebook variables.")

## 13. Global analysis of the `transactions` table

So far we had analyzed only a sample corresponding to a single daily `transactions` file.

In this section we'll go through **all the CSV files of the `transactions` table** available in S3 to get a global view of its main variables, without loading all ~5 million records into memory at once.

The goal is to calculate:

- Total number of transactions.
- Count and percentage of transactions flagged as fraud.
- Distribution by transaction type.
- Distribution by category.
- Channels used.
- Transaction status.
- Countries where they occurred.
- Merchant categories.
- Currencies used.

This will let us evaluate whether the dataset contains enough fraud cases and relevant patterns to proceed with a use case focused on **Fraud Detection**.

In [ ]:
from collections import Counter

# Categorical columns we want to summarize across the whole transactions table
columns_to_count = [
    "transaction_type",
    "transaction_category",
    "channel",
    "transaction_status",
    "is_fraud",
    "transaction_country",
    "merchant_category",
    "currency",
]

# One Counter per column
global_counts = {
    col: Counter()
    for col in columns_to_count
}

total_rows = 0
total_fraud = 0

# Filter only the CSV files for transactions
transaction_files = files_df[
    files_df["key"].str.startswith("data/transactions/")
    & files_df["key"].str.endswith(".csv")
]["key"].tolist()

print("Number of transactions files:", len(transaction_files))

for i, key in enumerate(transaction_files, start=1):
    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    # Only load the columns we care about
    df = pd.read_csv(
        obj["Body"],
        usecols=lambda col: col in columns_to_count
    )

    total_rows += len(df)

    # Accumulate counts
    for col in columns_to_count:
        if col in df.columns:
            values = df[col].fillna("NaN").value_counts()
            global_counts[col].update(values.to_dict())

    # Count total fraud
    if "is_fraud" in df.columns:
        total_fraud += (df["is_fraud"] == True).sum()

    if i % 100 == 0 or i == len(transaction_files):
        print(f"Processed {i}/{len(transaction_files)} files...")


print("\n==============================")
print("GLOBAL TRANSACTIONS SUMMARY")
print("==============================")

print(f"\nTotal records: {total_rows:,}")
print(f"Frauds found: {total_fraud:,}")

if total_rows > 0:
    fraud_rate = total_fraud / total_rows * 100
    print(f"Fraud rate: {fraud_rate:.4f}%")

for col in columns_to_count:
    print(f"\n### {col}")

    result_df = pd.DataFrame(
        global_counts[col].most_common(),
        columns=[col, "count"]
    )

    display(result_df.head(30))

Cantidad de archivos de transactions: 1097
Procesados 100/1097 archivos...
Procesados 200/1097 archivos...
Procesados 300/1097 archivos...
Procesados 400/1097 archivos...
Procesados 500/1097 archivos...
Procesados 600/1097 archivos...
Procesados 700/1097 archivos...
Procesados 800/1097 archivos...
Procesados 900/1097 archivos...
Procesados 1000/1097 archivos...
Procesados 1097/1097 archivos...

RESUMEN GLOBAL DE TRANSACTIONS

Total de registros: 4,425,008
Fraudes encontrados: 4,316
Tasa de fraude: 0.0975%

### transaction_type


,transaction_type,count
0,Purchase,1083406
1,Withdrawal,964673
2,Transfer,896438
3,Payment,738964
4,Deposit,609409
5,Adjustment,132118



### transaction_category


,transaction_category,count
0,NaN,2693520
1,Food,432468
2,Services,345370
3,Other,260518
4,Transport,260316
5,Entertainment,259643
6,Health,173173



### channel


,channel,count
0,POS,1548161
1,ATM,1328334
2,Web,663445
3,App,663414
4,Branch,132495
5,Transfer,89159



### transaction_status


,transaction_status,count
0,Approved,4070681
1,Declined,221234
2,Pending,88343
3,Reversed,44750



### is_fraud


,is_fraud,count
0,False,4420692
1,True,4316



### transaction_country


,transaction_country,count
0,México,2105794
1,Colombia,1289503
2,Argentina,867561
3,USA,40621
4,Spain,40542
5,Mexico,40515
6,Brazil,40472



### merchant_category


,merchant_category,count
0,NaN,3396215
1,Food,256846
2,Services,205124
3,Other,155029
4,Transport,154931
5,Entertainment,153960
6,Health,102903



### currency


,currency,count
0,USD,2437979
1,COP,1194444
2,ARS,792585


## 14. Fraud pattern analysis in `transactions`

The global analysis of the `transactions` table found:

- 4,425,008 transactions.
- 4,316 transactions flagged as fraud.
- An approximate fraud rate of 0.0975%.

Although fraud represents a very small proportion of the total, there are more than 4,000 positive cases, so there's enough information to investigate whether fraud-related patterns appear.

In this section we'll analyze how fraudulent transactions are distributed by:

- Transaction type.
- Channel.
- Country.
- Transaction category.
- Merchant category.
- Transaction status.
- Currency.
- Amount.

The goal is to discover whether fraud concentrates in certain types of operations or contexts, and to compare the fraud rate of each group, not just its absolute count.

In [17]:
from collections import defaultdict
import pandas as pd

dimensions = [
    "transaction_type",
    "channel",
    "transaction_country",
    "transaction_category",
    "merchant_category",
    "transaction_status",
    "currency",
]

# We'll store:
# total transactions and frauds for each value of each dimension
stats = {
    dimension: defaultdict(lambda: {"total": 0, "fraud": 0})
    for dimension in dimensions
}

# Amount statistics
fraud_amounts = []
non_fraud_amounts = []

transaction_files = files_df[
    files_df["key"].str.startswith("data/transactions/")
    & files_df["key"].str.endswith(".csv")
]["key"].tolist()

print(f"Transactions files to process: {len(transaction_files)}")

columns_needed = dimensions + [
    "is_fraud",
    "amount",
    "amount_usd",
]

for i, key in enumerate(transaction_files, start=1):

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    df = pd.read_csv(
        obj["Body"],
        usecols=lambda col: col in columns_needed
    )

    # -----------------------------------------
    # Counts per dimension
    # -----------------------------------------

    for dimension in dimensions:

        if dimension not in df.columns:
            continue

        temp = df[[dimension, "is_fraud"]].copy()

        temp[dimension] = temp[dimension].fillna("NaN")

        grouped = (
            temp.groupby(dimension)["is_fraud"]
            .agg(["count", "sum"])
        )

        for value, row in grouped.iterrows():

            stats[dimension][value]["total"] += int(row["count"])
            stats[dimension][value]["fraud"] += int(row["sum"])

    # -----------------------------------------
    # Amount distribution
    # -----------------------------------------

    amount_column = (
        "amount_usd"
        if "amount_usd" in df.columns
        else "amount"
    )

    if amount_column in df.columns:

        fraud_values = df.loc[
            df["is_fraud"] == True,
            amount_column
        ].dropna()

        non_fraud_values = df.loc[
            df["is_fraud"] == False,
            amount_column
        ].dropna()

        fraud_amounts.extend(fraud_values.tolist())

        # We don't need to store millions of normal amounts.
        # We take a small sample from each file.
        if len(non_fraud_values) > 0:
            sample_size = min(500, len(non_fraud_values))

            non_fraud_amounts.extend(
                non_fraud_values.sample(
                    sample_size,
                    random_state=42
                ).tolist()
            )

    if i % 100 == 0 or i == len(transaction_files):
        print(f"Processed {i}/{len(transaction_files)} files...")


print("\n======================================")
print("FRAUD PATTERNS BY DIMENSION")
print("======================================")

results = {}

for dimension in dimensions:

    rows = []

    for value, counts in stats[dimension].items():

        total = counts["total"]
        fraud = counts["fraud"]

        fraud_rate = (
            fraud / total * 100
            if total > 0
            else 0
        )

        rows.append({
            dimension: value,
            "total_transactions": total,
            "fraud_transactions": fraud,
            "fraud_rate_pct": fraud_rate
        })

    result_df = pd.DataFrame(rows)

    result_df = result_df.sort_values(
        "fraud_rate_pct",
        ascending=False
    )

    results[dimension] = result_df

    print(f"\n### {dimension}")

    display(
        result_df.head(30)
    )


print("\n======================================")
print("AMOUNT COMPARISON")
print("======================================")

amount_comparison = pd.DataFrame({
    "fraud": pd.Series(fraud_amounts).describe(),
    "non_fraud_sample": pd.Series(non_fraud_amounts).describe()
})

display(amount_comparison)

Archivos de transactions a procesar: 1097
Procesados 100/1097 archivos...
Procesados 200/1097 archivos...
Procesados 300/1097 archivos...
Procesados 400/1097 archivos...
Procesados 500/1097 archivos...
Procesados 600/1097 archivos...
Procesados 700/1097 archivos...
Procesados 800/1097 archivos...
Procesados 900/1097 archivos...
Procesados 1000/1097 archivos...
Procesados 1097/1097 archivos...

PATRONES DE FRAUDE POR DIMENSIÓN

### transaction_type


,transaction_type,total_transactions,fraud_transactions,fraud_rate_pct
0,Adjustment,132118,144,0.108993
3,Purchase,1083406,1086,0.100239
1,Deposit,609409,607,0.099605
2,Payment,738964,730,0.098787
5,Withdrawal,964673,951,0.098583
4,Transfer,896438,798,0.089019



### channel


,channel,total_transactions,fraud_transactions,fraud_rate_pct
2,Branch,132495,143,0.107929
0,ATM,1328334,1331,0.100201
5,Web,663445,656,0.098878
1,App,663414,634,0.095566
4,Transfer,89159,85,0.095335
3,POS,1548161,1467,0.094758



### transaction_country


,transaction_country,total_transactions,fraud_transactions,fraud_rate_pct
3,Mexico,40515,50,0.123411
1,Brazil,40472,45,0.111188
0,Argentina,867561,864,0.099590
4,México,2105794,2082,0.098870
5,Spain,40542,39,0.096197
6,USA,40621,38,0.093548
2,Colombia,1289503,1198,0.092904



### transaction_category


,transaction_category,total_transactions,fraud_transactions,fraud_rate_pct
2,Health,173173,185,0.106830
4,Other,260518,267,0.102488
5,Services,345370,341,0.098735
1,Food,432468,424,0.098042
0,Entertainment,259643,253,0.097441
3,NaN,2693520,2603,0.096639
6,Transport,260316,243,0.093348



### merchant_category


,merchant_category,total_transactions,fraud_transactions,fraud_rate_pct
2,Health,102903,111,0.107869
0,Entertainment,153960,163,0.105872
4,Other,155029,155,0.099981
1,Food,256846,254,0.098892
6,Transport,154931,150,0.096817
3,NaN,3396215,3287,0.096784
5,Services,205124,196,0.095552



### transaction_status


,transaction_status,total_transactions,fraud_transactions,fraud_rate_pct
0,Approved,4070681,3986,0.097920
1,Declined,221234,215,0.097182
2,Pending,88343,79,0.089424
3,Reversed,44750,36,0.080447



### currency


,currency,total_transactions,fraud_transactions,fraud_rate_pct
0,ARS,792585,787,0.099295
2,USD,2437979,2414,0.099016
1,COP,1194444,1115,0.093349



COMPARACIÓN DE MONTOS


,fraud,non_fraud_sample
count,1808.000000,548500.000000
mean,1664.573700,1676.920816
std,2319.124511,2336.287312
min,5.160000,5.000000
25%,247.720000,241.950000
50%,477.650000,466.610000
75%,1945.302500,1982.410000
max,9996.490000,9999.770000


## 15. Validating relationships between fraud, complaints, and calls

Before building an agent-assisted fraud investigation workflow, we need to verify whether the `transactions`, `complaints`, `call_transcripts`, and `call_center_interactions` tables can be reliably related to one another.

In this section we'll inspect a small sample of each table to identify:

- Shared keys such as `customer_id`, `account_id`, `transaction_id`, or `interaction_id`.
- Date fields that allow relating events close in time.
- The join path needed to connect a fraudulent transaction with complaints or calls from the same customer.

The subsequent goal will be to measure how many of the **4,316 frauds** have an associated complaint or call within time windows such as **±7, ±30, and ±90 days**.

This validation will let us decide whether it makes sense to build a **Fraud Investigation Agent** that combines transactional signals with NLP over complaints and transcripts.

In [ ]:
tables_to_inspect = [
    "transactions",
    "complaints",
    "call_transcripts",
    "call_center_interactions",
]

for table in tables_to_inspect:
    table_files = files_df[
        files_df["key"].str.startswith(f"data/{table}/")
        & files_df["key"].str.endswith(".csv")
    ]["key"].tolist()

    print("\n" + "=" * 80)
    print(f"TABLE: {table}")
    print("=" * 80)
    print(f"Number of files: {len(table_files)}")

    if not table_files:
        print("No files found.")
        continue

    sample_key = table_files[0]

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=sample_key,
    )

    sample_df = pd.read_csv(
        obj["Body"],
        nrows=5,
    )

    print("\nExample file:")
    print(sample_key)

    print("\nColumns:")
    for col in sample_df.columns:
        print(f"- {col}")

    print("\nSample:")
    display(sample_df)


TABLE: transactions
Cantidad de archivos: 1097

Archivo de ejemplo:
data/transactions/year=2023/month=06/day=17/transactions_20230617.csv

Columnas:
- transaction_id
- transaction_date
- process_date
- product_id
- customer_id
- transaction_type
- transaction_category
- amount
- currency
- amount_usd
- channel
- branch_id
- merchant_name
- merchant_category
- transaction_country
- transaction_city
- transaction_status
- response_code
- is_fraud
- fraud_score
- latitude
- longitude

Muestra:


,transaction_id,transaction_date,process_date,product_id,customer_id,transaction_type,transaction_category,amount,currency,amount_usd,...,merchant_name,merchant_category,transaction_country,transaction_city,transaction_status,response_code,is_fraud,fraud_score,latitude,longitude
0,TRX-7I07NJ7LT0TPC5YC33UL,2023-06-17 14:54:58,2023-06-17,PRD-7EFOVFRUE3UQ,CLI-DRS40C485MAB,Deposit,NaN,3695.53,USD,NaN,...,NaN,NaN,México,Monterrey,Approved,0,False,0.89,NaN,NaN
1,TRX-5COPO2MJ8G9XO7BR2TSE,2023-06-17 14:17:51,2023-06-17,PRD-HMGUZQXKE413,CLI-RIO7AY0XG4MY,Withdrawal,NaN,492.91,USD,NaN,...,NaN,NaN,México,Tijuana,Approved,0,False,25.02,NaN,NaN
2,TRX-SF2DZJYU0POJ9N4FGYV9,2023-06-18 03:17:02,2023-06-17,PRD-Z9VH3WRVEJP5,CLI-NRCMAC93SWIH,Withdrawal,NaN,1200383.22,COP,300.10,...,NaN,NaN,Argentina,Mendoza,Approved,0,False,NaN,NaN,NaN
3,TRX-VY3H2RR4HJ9VF7JQ7BZN,2023-06-18 03:48:24,2023-06-17,PRD-3963712YA073,CLI-NRS8RG7UHLKI,Withdrawal,NaN,10272.09,ARS,29.35,...,NaN,NaN,Argentina,Buenos Aires,Approved,0,False,1.74,NaN,NaN
4,TRX-8HE3NCEV1NN2QS36J56G,2023-06-17 07:08:37,2023-06-17,PRD-S6CRGMNKOGAT,CLI-M1CZMRQ8CW3G,Transfer,NaN,801.49,USD,NaN,...,NaN,NaN,México,Guadalajara,Approved,0,False,NaN,NaN,NaN



TABLE: complaints
Cantidad de archivos: 1097

Archivo de ejemplo:
data/complaints/year=2023/month=06/day=17/complaints_20230617.csv

Columnas:
- complaint_id
- creation_date
- process_date
- customer_id
- case_type
- category
- subcategory
- reception_channel
- affected_product_id
- related_branch_id
- origin_interaction_id
- description
- claimed_amount
- currency
- priority
- status
- assigned_agent_id
- assignment_date
- first_response_date
- resolution_date
- closing_date
- sla_breached
- resolution_days
- resolution
- compensation_granted
- resolution_satisfaction
- is_repeat_complainer

Muestra:


,complaint_id,creation_date,process_date,customer_id,case_type,category,subcategory,reception_channel,affected_product_id,related_branch_id,...,assignment_date,first_response_date,resolution_date,closing_date,sla_breached,resolution_days,resolution,compensation_granted,resolution_satisfaction,is_repeat_complainer
0,CMP-J7LT0TPC5YC33ULTQJZD,2023-06-17 16:54:58,2023-06-17,CLI-6TXMFGPFFTZG,Complaint,Transactions,Cargo no reconocido,Email,NaN,SUC-H24HUACK,...,NaN,NaN,NaN,NaN,True,NaN,NaN,NaN,NaN,False
1,CMP-P7HVN55QJQFLDZ4HRU55,2023-06-17 15:37:40,2023-06-17,CLI-PFYINQ8JMS43,Claim,Transactions,Cargo no reconocido,Email,PRD-ML7YZHXJWUCA,SUC-DG6C9C1K,...,2023-06-17 23:37:40,2023-06-18 03:37:40,2023-07-13 15:37:40,NaN,False,26.0,Se brindó explicación detallada al cliente y s...,457.44,NaN,False
2,CMP-JSJ0B0XVTB2VY3H2RR4H,2023-06-17 08:07:05,2023-06-17,CLI-5B1TIWCXY0U0,Suggestion,Technical,Problema con app,Branch,NaN,NaN,...,2023-06-18 04:07:05,2023-06-18 18:07:05,NaN,NaN,True,NaN,NaN,NaN,NaN,False
3,CMP-74XKHST63FFGQOZ3ECV8,2023-06-18 05:25:10,2023-06-17,CLI-7D9HFKZPAO82,Complaint,Branch,Atención en sucursal,Call Center,PRD-6W9PH9FVWZP4,SUC-JGB6PK8R,...,NaN,NaN,NaN,NaN,False,NaN,NaN,NaN,NaN,False
4,CMP-72Q2B1HXJ2IU22VEXKK1,2023-06-18 01:40:53,2023-06-17,CLI-737470INRZYR,Complaint,Service,NaN,App,PRD-84KOGPJBV6A5,NaN,...,2023-06-18 23:40:53,2023-06-20 22:40:53,NaN,NaN,False,NaN,NaN,NaN,NaN,False



TABLE: call_transcripts
Cantidad de archivos: 1097

Archivo de ejemplo:
data/call_transcripts/year=2023/month=06/day=17/call_transcripts_20230617.csv

Columnas:
- transcript_id
- interaction_id
- process_date
- customer_id
- agent_id
- full_text
- customer_text
- agent_text
- detected_language
- detected_accent
- accent_confidence
- detected_keywords
- mentioned_entities
- detected_intents
- main_topics
- transcription_model
- audio_quality
- duration_seconds

Muestra:


,transcript_id,interaction_id,process_date,customer_id,agent_id,full_text,customer_text,agent_text,detected_language,detected_accent,accent_confidence,detected_keywords,mentioned_entities,detected_intents,main_topics,transcription_model,audio_quality,duration_seconds
0,TRS-PPFGAVR5FA7R3DCC91G4,INT-R2TSE2WKP7HVN55Q,2023-06-17,CLI-C46C63F54F0A,AGT-7CQYEF4L7J,"Cliente: Hola, buenos días. Quisiera saber cuá...","Hola, buenos días. Quisiera saber cuál es mi s...","Buenos días, con gusto le ayudo. Permítame un ...",es,mexican,NaN,"banco, cuenta, servicio","{""account_numbers"": 0, ""dates"": 0, ""amounts"": ...",consulta_general,Comercial,Whisper v3,High,545.0
1,TRS-DD79XNL6Q5DZNKOUR81B,INT-ERY0VKHXAEWUJQSW,2023-06-17,CLI-RTCOQFG4LA8Q,AGT-67FCNFADT1,"Cliente: Hola, buenos días. Quisiera saber cuá...","Hola, buenos días. Quisiera saber cuál es mi s...","Buenos días, con gusto le ayudo. Permítame un ...",es,argentine,0.76,"cuenta, banco","{""account_numbers"": 0, ""dates"": 0, ""amounts"": ...",consulta_general,Queja,AWS Transcribe,High,470.0
2,TRS-E7ZI313MSF2DZJYU0POJ,INT-HC9ALVKI68EJ8HPG,2023-06-17,CLI-N344OED0JO0L,AGT-LTX82QBP86,"Cliente: Buenas tardes, necesito consultar el ...","Buenas tardes, necesito consultar el saldo de ...","Buenas tardes, claro que sí. Déjeme revisar es...",es,NaN,0.90,"cuenta, banco","{""account_numbers"": 0, ""dates"": 1, ""amounts"": ...",consulta_general,Queja,AWS Transcribe,Medium,420.0
3,TRS-H4DZT88K2QTZJ4NY5ILY,INT-AP0TE0VSZWQTWR8S,2023-06-17,CLI-XH0KSNXXSM0P,AGT-O8KA4K5V2R,"Cliente: Buenas tardes, necesito consultar el ...","Buenas tardes, necesito consultar el saldo de ...","Buenas tardes, claro que sí. Déjeme revisar es...",es,mexican,0.93,"banco, cuenta, servicio","{""account_numbers"": 0, ""dates"": 1, ""amounts"": ...",consulta_general,Técnico,Whisper v3,High,NaN
4,TRS-0P221BIAHJQVK8HE3NCE,INT-5X2T33QTKLKB4CNR,2023-06-17,CLI-FFFXXSRASE9S,AGT-HQP1WC8P45,"Cliente: Buenas tardes, necesito consultar el ...","Buenas tardes, necesito consultar el saldo de ...","Buenas tardes, claro que sí. Déjeme revisar es...",es,colombian,NaN,"banco, servicio","{""account_numbers"": 0, ""dates"": 1, ""amounts"": ...",consulta_general,Producto,Google STT,High,198.0



TABLE: call_center_interactions
Cantidad de archivos: 1097

Archivo de ejemplo:
data/call_center_interactions/year=2023/month=06/day=17/call_center_interactions_20230617.csv

Columnas:
- interaction_id
- interaction_date
- process_date
- customer_id
- agent_id
- interaction_type
- channel
- contact_reason
- reason_category
- duration_seconds
- wait_time_seconds
- was_resolved
- requires_followup
- detected_sentiment
- sentiment_score
- customer_detected_accent
- agent_used_accent
- was_escalated
- mentioned_products
- has_transcript
- has_recording

Muestra:


,interaction_id,interaction_date,process_date,customer_id,agent_id,interaction_type,channel,contact_reason,reason_category,duration_seconds,...,was_resolved,requires_followup,detected_sentiment,sentiment_score,customer_detected_accent,agent_used_accent,was_escalated,mentioned_products,has_transcript,has_recording
0,INT-8LKPPFGAVR5FA7R3,2023-06-17 18:00:48,2023-06-17,CLI-EBVG034XZPFT,AGT-HHYX8OE9JF,Inbound Call,Phone,Transaccional,Transaccional,250.0,...,True,False,Neutral,-0.28,argentine,argentine,False,NaN,False,True
1,INT-R2TSE2WKP7HVN55Q,2023-06-18 06:10:40,2023-06-17,CLI-C46C63F54F0A,AGT-7CQYEF4L7J,Inbound Call,Phone,Comercial,Comercial,545.0,...,True,False,Neutral,-0.17,mexican,mexican,False,"PRD-LWE7ZI313MSF,PRD-2DZJYU0POJ9N,PRD-4FGYV9I6...",True,True
2,INT-88K2QTZJ4NY5ILYX,2023-06-18 03:38:17,2023-06-17,CLI-ZOGX29P6Q6JB,AGT-3SCBQBHAUJ,Email,Email,Retención,Retención,NaN,...,True,False,Neutral,0.29,mexican,colombian,False,NaN,False,False
3,INT-3NJ03R8EHVW50EXS,2023-06-17 17:03:11,2023-06-17,CLI-XSM2WVAZ48IT,AGT-QZBFN7LRWT,Inbound Call,Phone,Comercial,Comercial,605.0,...,True,False,Muy Negativo,-0.76,argentine,mexican,False,NaN,False,True
4,INT-VD5UPMZN0TANX443,2023-06-17 08:20:20,2023-06-17,CLI-X00N9TOHD534,AGT-1SD8N8Y4VV,Inbound Call,Phone,Queja,Queja,603.0,...,True,False,Positivo,0.65,NaN,NaN,False,NaN,False,True


## 16. Relationship between fraudulent transactions, complaints, and calls

So far we identified 4,316 transactions flagged as fraud.

The next step is to determine whether those fraudulent transactions are temporally related to later or nearby interactions from the same customer.

We'll analyze whether a fraudulent transaction has:

- A complaint from the same customer within ±7 days.
- A complaint within ±30 days.
- A complaint within ±90 days.
- A call center interaction within those same windows.
- Either signal: call or complaint.

The relationship will be made via `customer_id`.

For dates we'll use:

- `transactions.transaction_date`
- `complaints.creation_date`
- `call_center_interactions.interaction_date`

The goal is to determine whether there's enough evidence to build a fraud investigation workflow that combines transactional signals with contact center information and, later, NLP over `call_transcripts`.

In [18]:
import pandas as pd

# ============================================================
# 1. LOAD FRAUDULENT TRANSACTIONS
# ============================================================

transaction_files = files_df[
    files_df["key"].str.startswith("data/transactions/")
    & files_df["key"].str.endswith(".csv")
]["key"].tolist()

fraud_parts = []

print("Loading fraudulent transactions...")

for i, key in enumerate(transaction_files, start=1):

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    df = pd.read_csv(
        obj["Body"],
        usecols=[
            "transaction_id",
            "customer_id",
            "transaction_date",
            "is_fraud",
        ]
    )

    fraud_mask = (
        (df["is_fraud"] == True)
        | df["is_fraud"].astype(str).str.lower().eq("true")
    )

    fraud_rows = df.loc[
        fraud_mask,
        ["transaction_id", "customer_id", "transaction_date"]
    ]

    if not fraud_rows.empty:
        fraud_parts.append(fraud_rows)

    if i % 200 == 0 or i == len(transaction_files):
        print(f"Transactions: {i}/{len(transaction_files)}")


fraud_df = pd.concat(
    fraud_parts,
    ignore_index=True
)

fraud_df["transaction_date"] = pd.to_datetime(
    fraud_df["transaction_date"],
    errors="coerce"
)

print(f"\nFrauds loaded: {len(fraud_df):,}")


# ============================================================
# 2. LOAD COMPLAINTS
# ============================================================

complaint_files = files_df[
    files_df["key"].str.startswith("data/complaints/")
    & files_df["key"].str.endswith(".csv")
]["key"].tolist()

complaint_parts = []

print("\nLoading complaints...")

for i, key in enumerate(complaint_files, start=1):

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    df = pd.read_csv(
        obj["Body"],
        usecols=[
            "complaint_id",
            "customer_id",
            "creation_date",
            "category",
            "subcategory",
            "description",
        ]
    )

    complaint_parts.append(df)

    if i % 200 == 0 or i == len(complaint_files):
        print(f"Complaints: {i}/{len(complaint_files)}")


complaints_df = pd.concat(
    complaint_parts,
    ignore_index=True
)

complaints_df["creation_date"] = pd.to_datetime(
    complaints_df["creation_date"],
    errors="coerce"
)

print(f"Complaints loaded: {len(complaints_df):,}")


# ============================================================
# 3. LOAD CALL CENTER INTERACTIONS
# ============================================================

interaction_files = files_df[
    files_df["key"].str.startswith("data/call_center_interactions/")
    & files_df["key"].str.endswith(".csv")
]["key"].tolist()

interaction_parts = []

print("\nLoading call center interactions...")

for i, key in enumerate(interaction_files, start=1):

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    df = pd.read_csv(
        obj["Body"],
        usecols=[
            "interaction_id",
            "customer_id",
            "interaction_date",
            "contact_reason",
            "reason_category",
            "detected_sentiment",
            "has_transcript",
        ]
    )

    interaction_parts.append(df)

    if i % 200 == 0 or i == len(interaction_files):
        print(
            f"Call center interactions: "
            f"{i}/{len(interaction_files)}"
        )


interactions_df = pd.concat(
    interaction_parts,
    ignore_index=True
)

interactions_df["interaction_date"] = pd.to_datetime(
    interactions_df["interaction_date"],
    errors="coerce"
)

print(
    f"Call center interactions loaded: "
    f"{len(interactions_df):,}"
)


# ============================================================
# 4. RELATE FRAUDS WITH COMPLAINTS
# ============================================================

print("\nRelating frauds with complaints...")

fraud_complaints = fraud_df.merge(
    complaints_df,
    on="customer_id",
    how="left"
)

fraud_complaints["days_from_fraud"] = (
    fraud_complaints["creation_date"]
    - fraud_complaints["transaction_date"]
).dt.total_seconds() / 86400


# ============================================================
# 5. RELATE FRAUDS WITH CALL CENTER
# ============================================================

print("Relating frauds with call center interactions...")

fraud_calls = fraud_df.merge(
    interactions_df,
    on="customer_id",
    how="left"
)

fraud_calls["days_from_fraud"] = (
    fraud_calls["interaction_date"]
    - fraud_calls["transaction_date"]
).dt.total_seconds() / 86400


# ============================================================
# 6. ANALYZE 7, 30, AND 90 DAY WINDOWS
# ============================================================

windows = [7, 30, 90]

results = []

total_frauds = fraud_df["transaction_id"].nunique()

for days in windows:

    complaints_in_window = fraud_complaints[
        fraud_complaints["days_from_fraud"].abs() <= days
    ]

    calls_in_window = fraud_calls[
        fraud_calls["days_from_fraud"].abs() <= days
    ]

    fraud_ids_with_complaint = set(
        complaints_in_window["transaction_id"]
    )

    fraud_ids_with_call = set(
        calls_in_window["transaction_id"]
    )

    fraud_ids_with_any = (
        fraud_ids_with_complaint
        | fraud_ids_with_call
    )

    results.append({
        "window_days": days,

        "total_frauds":
            total_frauds,

        "frauds_with_complaint":
            len(fraud_ids_with_complaint),

        "complaint_pct":
            len(fraud_ids_with_complaint)
            / total_frauds * 100,

        "frauds_with_call":
            len(fraud_ids_with_call),

        "call_pct":
            len(fraud_ids_with_call)
            / total_frauds * 100,

        "frauds_with_call_or_complaint":
            len(fraud_ids_with_any),

        "any_signal_pct":
            len(fraud_ids_with_any)
            / total_frauds * 100,
    })


# ============================================================
# 7. SHOW RESULTS
# ============================================================

relationship_summary = pd.DataFrame(results)

print("\n" + "=" * 60)
print("RELATIONSHIP BETWEEN FRAUD, COMPLAINTS, AND CALLS")
print("=" * 60)

display(relationship_summary)


# ============================================================
# 8. SHOW SOME EXAMPLES
# ============================================================

print("\nExamples of complaints near fraud:")

display(
    fraud_complaints[
        fraud_complaints["days_from_fraud"].abs() <= 30
    ][
        [
            "transaction_id",
            "customer_id",
            "transaction_date",
            "creation_date",
            "days_from_fraud",
            "category",
            "subcategory",
            "description",
        ]
    ].head(20)
)


print("\nExamples of calls near fraud:")

display(
    fraud_calls[
        fraud_calls["days_from_fraud"].abs() <= 30
    ][
        [
            "transaction_id",
            "customer_id",
            "transaction_date",
            "interaction_date",
            "days_from_fraud",
            "contact_reason",
            "reason_category",
            "detected_sentiment",
            "has_transcript",
        ]
    ].head(20)
)

Cargando transacciones fraudulentas...
Transactions: 200/1097
Transactions: 400/1097
Transactions: 600/1097
Transactions: 800/1097
Transactions: 1000/1097
Transactions: 1097/1097

Fraudes cargados: 4,316

Cargando complaints...
Complaints: 200/1097
Complaints: 400/1097
Complaints: 600/1097
Complaints: 800/1097
Complaints: 1000/1097
Complaints: 1097/1097
Complaints cargados: 67,095

Cargando call center interactions...
Call center interactions: 200/1097
Call center interactions: 400/1097
Call center interactions: 600/1097
Call center interactions: 800/1097
Call center interactions: 1000/1097
Call center interactions: 1097/1097
Call center interactions cargadas: 686,296

Relacionando fraudes con complaints...
Relacionando fraudes con call center interactions...

RELACIÓN ENTRE FRAUDE, QUEJAS Y LLAMADAS


,window_days,total_frauds,frauds_with_complaint,complaint_pct,frauds_with_call,call_pct,frauds_with_call_or_complaint,any_signal_pct
0,7,4316,19,0.440222,256,5.931418,274,6.348471
1,30,4316,80,1.853568,957,22.173309,1024,23.725672
2,90,4316,254,5.885079,2200,50.973123,2324,53.846154



Ejemplos de complaints cercanos a fraude:


,transaction_id,customer_id,transaction_date,creation_date,days_from_fraud,category,subcategory,description
187,TRX-1PRHLHU5CGWFFCO7LD92,CLI-XFIPH27VFSX2,2023-07-26 06:13:56,2023-07-15 08:46:06,-10.894329,Branch,Atención en sucursal,Queja relacionada con branch
209,TRX-MBA2AM8NBARCX1CKK1QZ,CLI-TYUNBA77B42H,2023-08-02 00:18:40,2023-07-19 19:32:29,-13.198738,Service,Calidad de servicio,Queja relacionada con service
267,TRX-JVJLEPTDJ0VRZWEC3AUV,CLI-USVEC3ISYK2G,2023-08-14 08:29:02,2023-08-13 17:44:33,-0.614225,Fees,Cobro indebido,Queja relacionada con fees
303,TRX-STJ4V0C9H3SBXNTJIIJP,CLI-ID2GCQ86PS16,2023-08-19 05:40:04,2023-09-05 11:38:18,17.248773,Fees,Cobro indebido,Queja relacionada con fees
364,TRX-4UO5DYVUQNE59LO4GM3T,CLI-MZ239VRK9DTK,2023-09-05 02:31:11,2023-10-01 21:25:34,26.787766,Technical,Problema con app,Queja relacionada con technical
519,TRX-R5H1XWQPOJKK5JUE2IIG,CLI-9JQL2ASY1UPD,2023-10-12 03:11:38,2023-09-22 22:09:17,-19.209965,Fees,Cobro indebido,Queja relacionada con fees
528,TRX-PVLZC3T39JQ9MGYQCM7R,CLI-BD422R9VJ0DV,2023-10-13 08:36:36,2023-10-26 06:32:15,12.913646,Fees,Cobro indebido,Queja relacionada con fees
588,TRX-QZI5NPEYLW2GW2TUF18M,CLI-R0QAKHQ2MYV8,2023-10-24 22:43:42,2023-10-15 19:41:01,-9.126863,Service,Calidad de servicio,Queja relacionada con service
621,TRX-NCFIQJMXMDYDXCEIBGNU,CLI-W7MF4YJ3IQV4,2023-11-01 21:19:29,2023-10-24 12:04:28,-8.385428,Branch,Atención en sucursal,Queja relacionada con branch
643,TRX-7S1EG1QH8ECBOA4OUZZY,CLI-OHEUKZ9TX6BQ,2023-11-07 21:47:11,2023-11-04 02:02:12,-3.822905,Transactions,NaN,Queja relacionada con transactions



Ejemplos de llamadas cercanas a fraude:


,transaction_id,customer_id,transaction_date,interaction_date,days_from_fraud,contact_reason,reason_category,detected_sentiment,has_transcript
21,TRX-6ABBHYXRSRF32TUXD5Y7,CLI-LPI5VHAVH21N,2023-06-20 04:27:29,2023-07-15 02:52:01,24.933704,Queja,Queja,Positivo,False
113,TRX-1MMALP0Y80V3CG0C3EIM,CLI-YUVL9N8IYDL9,2023-06-23 10:32:23,2023-07-17 10:26:40,23.996030,Transaccional,Transaccional,Neutral,False
198,TRX-M42NXYUKONJBFFT2WYMN,CLI-333F1ETY6YYP,2023-06-29 03:40:21,2023-07-25 04:12:04,26.022025,Transaccional,Transaccional,Neutral,False
203,TRX-P5ERVJYUMCXYE7YFZ58Z,CLI-2HYLAJQ5HOQ6,2023-06-28 19:35:35,2023-06-28 09:01:46,-0.440150,Queja,Queja,Neutral,False
206,TRX-TVX0KJQEDDVLKVKTB145,CLI-5IN6TYXMWI99,2023-06-28 19:02:34,2023-06-22 01:26:20,-6.733495,Transaccional,Transaccional,Neutral,False
263,TRX-NIE77S1982JSSDEHLRUT,CLI-895G0OPN8DN4,2023-07-01 18:25:02,2023-07-31 09:13:33,29.617025,Técnico,Técnico,Positivo,False
290,TRX-FDFY10Q5R40HZAQ1PQ1K,CLI-0SSVWFL67CIR,2023-07-03 13:29:53,2023-06-25 23:14:40,-7.593900,Queja,Queja,Neutral,False
373,TRX-G1GOZDOVFSTVH8YD7BTP,CLI-AYUNN4O9VU5K,2023-07-06 21:57:47,2023-08-04 04:52:53,28.288264,Técnico,Técnico,Negativo,False
384,TRX-N23RJ4Q0B782CYBJVQ6X,CLI-QE0NVDI14KA0,2023-07-07 05:19:46,2023-07-17 00:43:39,9.808252,Transaccional,Transaccional,Neutral,False
419,TRX-R4VYZ1UJ6TTVL44EHZWS,CLI-K2DR3G10UWAK,2023-07-09 03:57:10,2023-07-12 03:58:06,3.000648,Técnico,Técnico,Positivo,False


## 17. Analysis of `call_transcripts` associated with fraud

In the previous section we found that:

- 957 fraudulent transactions have at least one call center interaction within ±30 days.
- 2,200 fraudulent transactions have at least one interaction within ±90 days.

However, an interaction close to a fraud event doesn't necessarily mean the customer talked about the fraud.

In this section we'll answer two questions:

1. How many of the calls near a fraudulent transaction have a transcript available?
2. How many of those transcripts contain language potentially related to fraud?

For this we'll use:

- `call_center_interactions.interaction_id`
- `call_center_interactions.has_transcript`
- `call_transcripts.interaction_id`
- `call_transcripts.full_text`
- `call_transcripts.customer_text`
- `call_transcripts.detected_keywords`
- `call_transcripts.detected_intents`
- `call_transcripts.main_topics`

First we'll analyze a ±30-day window around each fraudulent transaction.

The fraud-related language detection will initially be a heuristic based on frequent terms and expressions such as:

- "I don't recognize"
- "I didn't do that"
- "unknown charge"
- "unauthorized transaction"
- "I was robbed"
- "fraud"
- "block my card"

This heuristic will serve as a first approximation or baseline. It can later be compared against a learned NLP or LLM-based component.

In [19]:
import pandas as pd
import re

# ============================================================
# 1. TAKE CALLS NEAR FRAUD WITHIN ±30 DAYS
# ============================================================

fraud_calls_30 = fraud_calls[
    fraud_calls["days_from_fraud"].abs() <= 30
].copy()

print("Fraud-call relationships within ±30 days:")
print(f"{len(fraud_calls_30):,}")

print("\nUnique frauds with a call within ±30 days:")
print(f"{fraud_calls_30['transaction_id'].nunique():,}")

print("\nUnique interactions:")
print(f"{fraud_calls_30['interaction_id'].nunique():,}")


# ============================================================
# 2. MEASURE HOW MANY INTERACTIONS HAVE A TRANSCRIPT
# ============================================================

unique_interactions_30 = (
    fraud_calls_30[
        [
            "interaction_id",
            "transaction_id",
            "customer_id",
            "interaction_date",
            "days_from_fraud",
            "has_transcript",
        ]
    ]
    .drop_duplicates("interaction_id")
)

interactions_with_transcript_flag = unique_interactions_30[
    unique_interactions_30["has_transcript"] == True
]

print("\n========================================")
print("AVAILABLE TRANSCRIPTS")
print("========================================")

print(
    f"Unique interactions within ±30 days: "
    f"{len(unique_interactions_30):,}"
)

print(
    f"Flagged with has_transcript=True: "
    f"{len(interactions_with_transcript_flag):,}"
)

if len(unique_interactions_30) > 0:
    pct = (
        len(interactions_with_transcript_flag)
        / len(unique_interactions_30)
        * 100
    )

    print(
        f"Percentage with transcript: "
        f"{pct:.2f}%"
    )


# ============================================================
# 3. LOAD CALL_TRANSCRIPTS
# ============================================================

transcript_files = files_df[
    files_df["key"].str.startswith("data/call_transcripts/")
    & files_df["key"].str.endswith(".csv")
]["key"].tolist()

print(
    f"\ncall_transcripts files found: "
    f"{len(transcript_files)}"
)

transcript_parts = []

for i, key in enumerate(transcript_files, start=1):

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    df = pd.read_csv(
        obj["Body"],
        usecols=[
            "transcript_id",
            "interaction_id",
            "customer_id",
            "full_text",
            "customer_text",
            "detected_keywords",
            "detected_intents",
            "main_topics",
        ]
    )

    transcript_parts.append(df)

    if i % 200 == 0 or i == len(transcript_files):
        print(
            f"Call transcripts: "
            f"{i}/{len(transcript_files)} files"
        )


transcripts_df = pd.concat(
    transcript_parts,
    ignore_index=True
)

print(
    f"\nTotal transcripts loaded: "
    f"{len(transcripts_df):,}"
)


# ============================================================
# 4. KEEP ONLY TRANSCRIPTS RELATED
#    TO CALLS NEAR FRAUD
# ============================================================

fraud_transcripts_30 = unique_interactions_30.merge(
    transcripts_df,
    on="interaction_id",
    how="inner",
    suffixes=("_interaction", "_transcript")
)

print("\n========================================")
print("FRAUDS WITH TRANSCRIPT")
print("========================================")

print(
    f"Transcripts associated with calls "
    f"near fraud: "
    f"{len(fraud_transcripts_30):,}"
)

frauds_with_real_transcript = (
    fraud_transcripts_30["transaction_id"]
    .nunique()
)

print(
    f"Unique frauds with at least one transcript: "
    f"{frauds_with_real_transcript:,}"
)

total_frauds_30 = (
    fraud_calls_30["transaction_id"]
    .nunique()
)

if total_frauds_30 > 0:
    fraud_transcript_pct = (
        frauds_with_real_transcript
        / total_frauds_30
        * 100
    )

    print(
        f"Percentage of frauds with a call "
        f"that have a transcript: "
        f"{fraud_transcript_pct:.2f}%"
    )


# ============================================================
# 5. CREATE UNIFIED TEXT FOR ANALYSIS
# ============================================================

text_columns = [
    "customer_text",
    "full_text",
    "detected_keywords",
    "detected_intents",
    "main_topics",
]

for col in text_columns:
    fraud_transcripts_30[col] = (
        fraud_transcripts_30[col]
        .fillna("")
        .astype(str)
    )

fraud_transcripts_30["combined_text"] = (
    fraud_transcripts_30[text_columns]
    .agg(" ".join, axis=1)
    .str.lower()
)


# ============================================================
# 6. HEURISTIC BASELINE:
#    SEARCH FOR FRAUD-RELATED LANGUAGE
# ============================================================

# NOTE: these patterns intentionally stay in Spanish — they match
# the actual customer-facing text in the dataset (LATAM bank), not
# notebook documentation.
fraud_patterns = [
    r"\bfraude\b",
    r"\bfraudul",
    r"no reconozco",
    r"no reconocido",
    r"no autoriz",
    r"yo no hice",
    r"no fui yo",
    r"cargo desconocido",
    r"compra desconocida",
    r"transacci[oó]n desconocida",
    r"transacci[oó]n no autorizada",
    r"movimiento desconocido",
    r"movimiento que no",
    r"me robaron",
    r"me robaron la tarjeta",
    r"tarjeta robada",
    r"perd[ií] mi tarjeta",
    r"tarjeta perdida",
    r"clonaron",
    r"clonaci[oó]n",
    r"suplantaci[oó]n",
    r"hackearon",
    r"hackeo",
    r"bloquear tarjeta",
    r"bloqueo de tarjeta",
]

fraud_regex = re.compile(
    "|".join(fraud_patterns),
    flags=re.IGNORECASE
)

fraud_transcripts_30["fraud_language_flag"] = (
    fraud_transcripts_30["combined_text"]
    .str.contains(
        fraud_regex,
        na=False
    )
)


# ============================================================
# 7. BASELINE NLP RESULTS
# ============================================================

fraud_language_transcripts = fraud_transcripts_30[
    fraud_transcripts_30["fraud_language_flag"]
]

frauds_with_fraud_language = (
    fraud_language_transcripts[
        "transaction_id"
    ]
    .nunique()
)

print("\n========================================")
print("FRAUD-RELATED LANGUAGE")
print("========================================")

print(
    f"Transcripts with fraud-related "
    f"language: "
    f"{len(fraud_language_transcripts):,}"
)

print(
    f"Unique frauds with at least one transcript "
    f"containing fraud-related language: "
    f"{frauds_with_fraud_language:,}"
)

if frauds_with_real_transcript > 0:
    language_pct = (
        frauds_with_fraud_language
        / frauds_with_real_transcript
        * 100
    )

    print(
        f"Percentage among frauds with a transcript: "
        f"{language_pct:.2f}%"
    )


# ============================================================
# 8. FINAL FUNNEL
# ============================================================

funnel = pd.DataFrame({
    "stage": [
        "Total frauds",
        "Frauds with call ±30 days",
        "Frauds with transcript",
        "Frauds with fraud-related language",
    ],
    "count": [
        fraud_df["transaction_id"].nunique(),
        total_frauds_30,
        frauds_with_real_transcript,
        frauds_with_fraud_language,
    ]
})

print("\n========================================")
print("FRAUD + CALL TRANSCRIPTS FUNNEL")
print("========================================")

display(funnel)


# ============================================================
# 9. SHOW EXAMPLES
# ============================================================

print(
    "\nExamples of transcripts detected "
    "by the heuristic:"
)

display(
    fraud_language_transcripts[
        [
            "transaction_id",
            "interaction_id",
            "days_from_fraud",
            "customer_text",
            "detected_keywords",
            "detected_intents",
            "main_topics",
        ]
    ].head(20)
)

Relaciones fraude-llamada en ±30 días:
1,092

Fraudes únicos con llamada en ±30 días:
957

Interacciones únicas:
1,091

TRANSCRIPTS DISPONIBLES
Interacciones únicas en ±30 días: 1,091
Marcadas con has_transcript=True: 256
Porcentaje con transcript: 23.46%

Archivos de call_transcripts encontrados: 1097
Call transcripts: 200/1097 archivos
Call transcripts: 400/1097 archivos
Call transcripts: 600/1097 archivos
Call transcripts: 800/1097 archivos
Call transcripts: 1000/1097 archivos
Call transcripts: 1097/1097 archivos

Total transcripts cargados: 171,321

FRAUDES CON TRANSCRIPT
Transcripts asociados a llamadas cercanas a fraude: 256
Fraudes únicos con al menos un transcript: 247
Porcentaje de fraudes con llamada que tienen transcript: 25.81%

LENGUAJE RELACIONADO CON FRAUDE
Transcripts con lenguaje relacionado con fraude: 0
Fraudes únicos con al menos un transcript con lenguaje relacionado con fraude: 0
Porcentaje entre los fraudes con transcript: 0.00%

FUNNEL FRAUDE + CALL TRANSCRIPTS


,stage,count
0,Fraudes totales,4316
1,Fraudes con llamada ±30 días,957
2,Fraudes con transcript,247
3,Fraudes con lenguaje relacionado con fraude,0



Ejemplos de transcripts detectados por la heurística:


,transaction_id,interaction_id,days_from_fraud,customer_text,detected_keywords,detected_intents,main_topics


## 18. Semantic exploration of fraud-related transcripts

In the previous section we found the following funnel:

- 4,316 fraudulent transactions.
- 957 frauds with at least one call within ±30 days.
- 247 frauds with at least one transcript available.
- 0 cases detected by a simple heuristic based on fraud-related keywords.

However, that result doesn't necessarily mean the transcripts don't contain useful signals.

The previous heuristic looked for explicit expressions such as:

- "I don't recognize"
- "I didn't do that"
- "fraud"
- "stolen card"
- "unauthorized transaction"

It's possible the real language in the dataset uses other expressions, intents, topics, or keywords.

In this section we'll explore the content of fraud-associated transcripts without imposing a classification yet.

We'll analyze:

- `detected_intents`
- `main_topics`
- `detected_keywords`
- `customer_text`
- `full_text`

The goal is to discover what language actually appears in these 256 transcripts and determine whether there's any useful semantic signal for building an NLP component.

In [20]:
from collections import Counter
import pandas as pd
import re

# ============================================================
# 1. GENERAL SUMMARY OF FRAUD-RELATED TRANSCRIPTS
# ============================================================

print("=" * 70)
print("SEMANTIC EXPLORATION OF FRAUD-ASSOCIATED TRANSCRIPTS")
print("=" * 70)

print(f"\nTotal rows: {len(fraud_transcripts_30):,}")
print(
    "Unique transcripts:",
    fraud_transcripts_30["transcript_id"].nunique()
)

print(
    "Unique frauds represented:",
    fraud_transcripts_30["transaction_id"].nunique()
)


# ============================================================
# 2. ANALYZE DETECTED_INTENTS
# ============================================================

print("\n" + "=" * 70)
print("TOP DETECTED_INTENTS")
print("=" * 70)

intent_counter = Counter()

for value in fraud_transcripts_30["detected_intents"].dropna():

    text = str(value).strip()

    if not text:
        continue

    # Possible separators in the dataset
    parts = re.split(r"[,;|]", text)

    for part in parts:
        part = part.strip()

        if part:
            intent_counter[part] += 1


intents_df = pd.DataFrame(
    intent_counter.most_common(30),
    columns=[
        "detected_intent",
        "count"
    ]
)

display(intents_df)


# ============================================================
# 3. ANALYZE MAIN_TOPICS
# ============================================================

print("\n" + "=" * 70)
print("TOP MAIN_TOPICS")
print("=" * 70)

topic_counter = Counter()

for value in fraud_transcripts_30["main_topics"].dropna():

    text = str(value).strip()

    if not text:
        continue

    parts = re.split(r"[,;|]", text)

    for part in parts:
        part = part.strip()

        if part:
            topic_counter[part] += 1


topics_df = pd.DataFrame(
    topic_counter.most_common(30),
    columns=[
        "main_topic",
        "count"
    ]
)

display(topics_df)


# ============================================================
# 4. ANALYZE DETECTED_KEYWORDS
# ============================================================

print("\n" + "=" * 70)
print("TOP DETECTED_KEYWORDS")
print("=" * 70)

keyword_counter = Counter()

for value in fraud_transcripts_30["detected_keywords"].dropna():

    text = str(value).strip()

    if not text:
        continue

    parts = re.split(r"[,;|]", text)

    for part in parts:
        part = part.strip().lower()

        if part:
            keyword_counter[part] += 1


keywords_df = pd.DataFrame(
    keyword_counter.most_common(50),
    columns=[
        "keyword",
        "count"
    ]
)

display(keywords_df)


# ============================================================
# 5. SHOW CUSTOMER_TEXT
# ============================================================

print("\n" + "=" * 70)
print("EXAMPLES OF CUSTOMER_TEXT")
print("=" * 70)

customer_text_examples = (
    fraud_transcripts_30[
        [
            "transaction_id",
            "interaction_id",
            "days_from_fraud",
            "customer_text",
            "detected_intents",
            "main_topics",
            "detected_keywords",
        ]
    ]
    .dropna(subset=["customer_text"])
)

display(
    customer_text_examples.head(30)
)


# ============================================================
# 6. SHOW FULL_TEXT
# ============================================================

print("\n" + "=" * 70)
print("EXAMPLES OF FULL_TEXT")
print("=" * 70)

full_text_examples = (
    fraud_transcripts_30[
        [
            "transaction_id",
            "interaction_id",
            "days_from_fraud",
            "full_text",
        ]
    ]
    .dropna(subset=["full_text"])
)

display(
    full_text_examples.head(20)
)


# ============================================================
# 7. MOST FREQUENT WORDS IN CUSTOMER_TEXT
# ============================================================

print("\n" + "=" * 70)
print("MOST FREQUENT WORDS IN CUSTOMER_TEXT")
print("=" * 70)

# NOTE: stopword list intentionally stays in Spanish — it filters
# actual Spanish customer-text DATA, it's not notebook documentation.
stopwords = {
    "de", "la", "el", "que", "y", "en", "a", "los", "las",
    "un", "una", "por", "para", "con", "del", "al", "mi",
    "me", "es", "no", "se", "lo", "le", "su", "ya", "pero",
    "porque", "como", "más", "muy", "esto", "esa", "ese",
    "quiero", "tengo", "puede", "podría"
}

word_counter = Counter()

for text in fraud_transcripts_30["customer_text"].dropna():

    words = re.findall(
        r"\b[a-záéíóúñü]{3,}\b",
        str(text).lower()
    )

    for word in words:

        if word not in stopwords:
            word_counter[word] += 1


words_df = pd.DataFrame(
    word_counter.most_common(50),
    columns=[
        "word",
        "count"
    ]
)

display(words_df)


# ============================================================
# 8. TRANSCRIPTS CLOSEST IN TIME TO THE FRAUD EVENT
# ============================================================

print("\n" + "=" * 70)
print("TRANSCRIPTS CLOSEST TO THE FRAUD EVENT")
print("=" * 70)

closest_transcripts = (
    fraud_transcripts_30
    .assign(
        abs_days_from_fraud=
        fraud_transcripts_30["days_from_fraud"].abs()
    )
    .sort_values("abs_days_from_fraud")
)

display(
    closest_transcripts[
        [
            "transaction_id",
            "interaction_id",
            "days_from_fraud",
            "customer_text",
            "detected_keywords",
            "detected_intents",
            "main_topics",
        ]
    ].head(30)
)


# ============================================================
# 9. SUMMARY OF AVAILABLE FIELDS
# ============================================================

summary = pd.DataFrame({
    "field": [
        "customer_text",
        "full_text",
        "detected_keywords",
        "detected_intents",
        "main_topics",
    ],
    "non_null_count": [
        fraud_transcripts_30["customer_text"].notna().sum(),
        fraud_transcripts_30["full_text"].notna().sum(),
        fraud_transcripts_30["detected_keywords"].notna().sum(),
        fraud_transcripts_30["detected_intents"].notna().sum(),
        fraud_transcripts_30["main_topics"].notna().sum(),
    ]
})

summary["pct_available"] = (
    summary["non_null_count"]
    / len(fraud_transcripts_30)
    * 100
)

print("\n" + "=" * 70)
print("FIELD AVAILABILITY")
print("=" * 70)

display(summary)

EXPLORACIÓN SEMÁNTICA DE TRANSCRIPTS ASOCIADOS A FRAUDE

Total de filas: 256
Transcripts únicos: 256
Fraudes únicos representados: 247

TOP DETECTED_INTENTS


,detected_intent,count
0,consulta_general,241



TOP MAIN_TOPICS


,main_topic,count
0,Transaccional,97
1,Producto,54
2,Técnico,43
3,Queja,38
4,Comercial,19
5,Retención,5



TOP DETECTED_KEYWORDS


,keyword,count
0,cuenta,217
1,servicio,216
2,banco,214



EJEMPLOS DE CUSTOMER_TEXT


,transaction_id,interaction_id,days_from_fraud,customer_text,detected_intents,main_topics,detected_keywords
0,TRX-EJSM8SXVVAADSD38APQA,INT-7NFRSU6CG4T3QANY,0.336933,"Buenas tardes, necesito consultar el saldo de ...",consulta_general,Transaccional,"cuenta, servicio"
1,TRX-EP43IHR8C16PNY2R80GV,INT-6AY38JMFB0W0X6TR,-24.965660,"Buenas tardes, necesito consultar el saldo de ...",consulta_general,Queja,"banco, cuenta"
2,TRX-LG2SFVJC75RBBMTDYJBL,INT-8OQ6GH2OHQ5GE1YV,-10.609792,"Buenas tardes, necesito consultar el saldo de ...",consulta_general,Transaccional,"banco, cuenta"
3,TRX-LG2SFVJC75RBBMTDYJBL,INT-1TG5V5953PFFD7OS,23.615197,"Buenas tardes, necesito consultar el saldo de ...",consulta_general,Producto,"cuenta, banco, servicio"
4,TRX-BRD8X48T9Y4FUGP0VPN5,INT-XPDQ3OU04LIAPL9J,1.340787,"Buenas tardes, necesito consultar el saldo de ...",consulta_general,Técnico,"cuenta, servicio, banco"
5,TRX-X1ZHBBETCHX1IVG16BJD,INT-THHH6U667Q6BOFHQ,13.519896,"Hola, buenos días. Quisiera saber cuál es mi s...",consulta_general,Transaccional,"cuenta, servicio"
6,TRX-5NNWZFC9AJC5ZGFKART6,INT-8L3KFIDNVVVLSLXM,10.648148,"Hola, buenos días. Quisiera saber cuál es mi s...",consulta_general,Transaccional,"cuenta, banco"
7,TRX-73XKCII1VBAYHZBKWL8W,INT-KJM9TV49I62RBNNV,-16.319201,"Hola, buenos días. Quisiera saber cuál es mi s...",consulta_general,Técnico,"servicio, cuenta, banco"
8,TRX-SVXW7T3XEW4VDYDXXG44,INT-7HG4QTZC7J0NQ6A5,3.019919,"Buenas tardes, necesito consultar el saldo de ...",,Transaccional,"servicio, cuenta, banco"
9,TRX-R58U1ZEGXDPY2QVOAZYA,INT-5HOQZZIR2W5S9OGC,-21.534086,"Buenas tardes, necesito consultar el saldo de ...",consulta_general,Producto,"servicio, banco"



EJEMPLOS DE FULL_TEXT


,transaction_id,interaction_id,days_from_fraud,full_text
0,TRX-EJSM8SXVVAADSD38APQA,INT-7NFRSU6CG4T3QANY,0.336933,"Cliente: Buenas tardes, necesito consultar el ..."
1,TRX-EP43IHR8C16PNY2R80GV,INT-6AY38JMFB0W0X6TR,-24.965660,"Cliente: Buenas tardes, necesito consultar el ..."
2,TRX-LG2SFVJC75RBBMTDYJBL,INT-8OQ6GH2OHQ5GE1YV,-10.609792,"Cliente: Buenas tardes, necesito consultar el ..."
3,TRX-LG2SFVJC75RBBMTDYJBL,INT-1TG5V5953PFFD7OS,23.615197,"Cliente: Buenas tardes, necesito consultar el ..."
4,TRX-BRD8X48T9Y4FUGP0VPN5,INT-XPDQ3OU04LIAPL9J,1.340787,"Cliente: Buenas tardes, necesito consultar el ..."
5,TRX-X1ZHBBETCHX1IVG16BJD,INT-THHH6U667Q6BOFHQ,13.519896,"Cliente: Hola, buenos días. Quisiera saber cuá..."
6,TRX-5NNWZFC9AJC5ZGFKART6,INT-8L3KFIDNVVVLSLXM,10.648148,"Cliente: Hola, buenos días. Quisiera saber cuá..."
7,TRX-73XKCII1VBAYHZBKWL8W,INT-KJM9TV49I62RBNNV,-16.319201,"Cliente: Hola, buenos días. Quisiera saber cuá..."
8,TRX-SVXW7T3XEW4VDYDXXG44,INT-7HG4QTZC7J0NQ6A5,3.019919,"Cliente: Buenas tardes, necesito consultar el ..."
9,TRX-R58U1ZEGXDPY2QVOAZYA,INT-5HOQZZIR2W5S9OGC,-21.534086,"Cliente: Buenas tardes, necesito consultar el ..."



PALABRAS MÁS FRECUENTES EN CUSTOMER_TEXT


,word,count
0,saldo,256
1,saber,177
2,buenas,130
3,tardes,130
4,necesito,130
5,consultar,130
6,tarjeta,130
7,crédito,130
8,hola,126
9,buenos,126



TRANSCRIPTS MÁS CERCANOS AL EVENTO DE FRAUDE


,transaction_id,interaction_id,days_from_fraud,customer_text,detected_keywords,detected_intents,main_topics
247,TRX-Y8J9UDWZP72EOZ5W35M1,INT-1GQKIKQYS32ODN4L,-0.189259,"Buenas tardes, necesito consultar el saldo de ...","banco, servicio",consulta_general,Queja
0,TRX-EJSM8SXVVAADSD38APQA,INT-7NFRSU6CG4T3QANY,0.336933,"Buenas tardes, necesito consultar el saldo de ...","cuenta, servicio",consulta_general,Transaccional
35,TRX-QI2KJRGNQ3Q7EBWXB9PR,INT-LGRW72F7AQRU3XML,-0.360382,"Buenas tardes, necesito consultar el saldo de ...","servicio, banco, cuenta",consulta_general,Transaccional
68,TRX-NQ4IC3MRDRCRT5R3FVYY,INT-MGP8056PUGUZUAQH,0.365243,"Hola, buenos días. Quisiera saber cuál es mi s...","banco, cuenta",consulta_general,Transaccional
180,TRX-AJEP66GBETNTO9F0WY7J,INT-8A1DT404FIEKK7QU,-0.415590,"Hola, buenos días. Quisiera saber cuál es mi s...","cuenta, servicio, banco",consulta_general,Transaccional
252,TRX-IPPXO2K7QN3S0H3PM7HZ,INT-B7YF02LMWATWEF0K,-0.428588,"Hola, buenos días. Quisiera saber cuál es mi s...","cuenta, banco, servicio",consulta_general,Transaccional
19,TRX-CF9Y63SQ2KOKGLWFLUS8,INT-E9YHD6129AJCBA3N,0.492049,"Buenas tardes, necesito consultar el saldo de ...","cuenta, banco, servicio",consulta_general,Transaccional
95,TRX-N0NTRPDPK7LUK28DPH94,INT-8O4I479YA32EXB79,-0.628519,"Buenas tardes, necesito consultar el saldo de ...","servicio, cuenta, banco",consulta_general,Comercial
208,TRX-UPG05SVVGI2WRSSMD8N7,INT-X4U2UZYDXPC8M3K1,-0.807616,"Hola, buenos días. Quisiera saber cuál es mi s...","cuenta, banco",consulta_general,Transaccional
107,TRX-JQ12I4PT4WSJOFJPQDF9,INT-O4VRQ3GCZKKUA5EB,0.879028,"Buenas tardes, necesito consultar el saldo de ...","cuenta, banco, servicio",consulta_general,Transaccional



DISPONIBILIDAD DE CAMPOS


,field,non_null_count,pct_available
0,customer_text,256,100.0
1,full_text,256,100.0
2,detected_keywords,256,100.0
3,detected_intents,256,100.0
4,main_topics,256,100.0


## 19. Comparison of interactions near fraud vs. non-fraud

So far we've observed that transcripts temporally associated with fraud contain fairly generic text, and that the `detected_intents`, `detected_keywords`, and `main_topics` fields show no explicit fraud signals.

However, we still need to answer an important question:

> Do interactions near a fraudulent transaction behave differently from interactions near a non-fraudulent transaction?

In this section we'll compare a sample of fraudulent transactions against an equivalent sample of non-fraudulent transactions.

For both groups we'll analyze call center interactions within a ±30-day window and compare:

- Percentage of transactions with at least one interaction.
- Average number of interactions per transaction.
- Distribution of `reason_category`.
- Distribution of `detected_sentiment`.
- Transcript availability.
- Distribution of `main_topics` in the available transcripts.

The goal is to determine whether contact center interactions provide any differentiating signal between fraud and non-fraud.

If both populations show very similar distributions, that would indicate the contact center probably doesn't provide enough information for this fraud workflow in the current dataset.

In [21]:
import pandas as pd

# ============================================================
# 1. TAKE A SAMPLE OF NON-FRAUDULENT TRANSACTIONS
#    THE SAME SIZE AS THE FRAUD GROUP
# ============================================================

print("=" * 70)
print("FRAUD VS NON-FRAUD CONTACT CENTER COMPARISON")
print("=" * 70)

target_sample_size = fraud_df["transaction_id"].nunique()

non_fraud_parts = []

for i, key in enumerate(transaction_files, start=1):

    obj = s3.get_object(
        Bucket=BUCKET_NAME,
        Key=key
    )

    df = pd.read_csv(
        obj["Body"],
        usecols=[
            "transaction_id",
            "customer_id",
            "transaction_date",
            "is_fraud",
        ]
    )

    non_fraud_mask = (
        (df["is_fraud"] == False)
        | df["is_fraud"].astype(str).str.lower().eq("false")
    )

    non_fraud_rows = df.loc[
        non_fraud_mask,
        [
            "transaction_id",
            "customer_id",
            "transaction_date",
        ]
    ]

    if not non_fraud_rows.empty:
        non_fraud_parts.append(non_fraud_rows)

    if i % 200 == 0 or i == len(transaction_files):
        print(f"Transactions: {i}/{len(transaction_files)} files")


non_fraud_all = pd.concat(
    non_fraud_parts,
    ignore_index=True
)

non_fraud_all["transaction_date"] = pd.to_datetime(
    non_fraud_all["transaction_date"],
    errors="coerce"
)

non_fraud_sample = non_fraud_all.sample(
    n=target_sample_size,
    random_state=42
).copy()

print(
    f"\nFrauds to compare: "
    f"{target_sample_size:,}"
)

print(
    f"Sampled non-frauds: "
    f"{len(non_fraud_sample):,}"
)


# ============================================================
# 2. RELATE NON-FRAUDS WITH CALL CENTER
# ============================================================

non_fraud_calls = non_fraud_sample.merge(
    interactions_df,
    on="customer_id",
    how="left"
)

non_fraud_calls["days_from_transaction"] = (
    non_fraud_calls["interaction_date"]
    - non_fraud_calls["transaction_date"]
).dt.total_seconds() / 86400


# ============================================================
# 3. FILTER BOTH GROUPS TO ±30 DAYS
# ============================================================

fraud_calls_30 = fraud_calls[
    fraud_calls["days_from_fraud"].abs() <= 30
].copy()

non_fraud_calls_30 = non_fraud_calls[
    non_fraud_calls["days_from_transaction"].abs() <= 30
].copy()


# ============================================================
# 4. GENERAL INTERACTION METRICS
# ============================================================

fraud_transaction_count = (
    fraud_df["transaction_id"].nunique()
)

non_fraud_transaction_count = (
    non_fraud_sample["transaction_id"].nunique()
)

fraud_with_call = (
    fraud_calls_30["transaction_id"].nunique()
)

non_fraud_with_call = (
    non_fraud_calls_30["transaction_id"].nunique()
)


general_comparison = pd.DataFrame({
    "group": [
        "Fraud",
        "Non-Fraud",
    ],

    "transactions": [
        fraud_transaction_count,
        non_fraud_transaction_count,
    ],

    "transactions_with_call_30d": [
        fraud_with_call,
        non_fraud_with_call,
    ],

    "pct_with_call_30d": [
        fraud_with_call
        / fraud_transaction_count * 100,

        non_fraud_with_call
        / non_fraud_transaction_count * 100,
    ],

    "interaction_rows_30d": [
        len(fraud_calls_30),
        len(non_fraud_calls_30),
    ],

    "avg_interactions_per_transaction": [
        len(fraud_calls_30)
        / fraud_transaction_count,

        len(non_fraud_calls_30)
        / non_fraud_transaction_count,
    ],
})

print("\n" + "=" * 70)
print("GENERAL COMPARISON")
print("=" * 70)

display(general_comparison)


# ============================================================
# 5. FUNCTION TO COMPARE DISTRIBUTIONS
# ============================================================

def compare_distribution(
    fraud_data,
    non_fraud_data,
    column
):

    fraud_counts = (
        fraud_data[column]
        .fillna("NaN")
        .value_counts()
        .rename("fraud_count")
    )

    non_fraud_counts = (
        non_fraud_data[column]
        .fillna("NaN")
        .value_counts()
        .rename("non_fraud_count")
    )

    comparison = pd.concat(
        [
            fraud_counts,
            non_fraud_counts
        ],
        axis=1
    ).fillna(0)

    comparison["fraud_pct"] = (
        comparison["fraud_count"]
        / comparison["fraud_count"].sum()
        * 100
    )

    comparison["non_fraud_pct"] = (
        comparison["non_fraud_count"]
        / comparison["non_fraud_count"].sum()
        * 100
    )

    comparison["difference_pct_points"] = (
        comparison["fraud_pct"]
        - comparison["non_fraud_pct"]
    )

    return comparison.sort_values(
        "difference_pct_points",
        ascending=False
    )


# ============================================================
# 6. COMPARE REASON_CATEGORY
# ============================================================

print("\n" + "=" * 70)
print("REASON_CATEGORY")
print("=" * 70)

reason_comparison = compare_distribution(
    fraud_calls_30,
    non_fraud_calls_30,
    "reason_category"
)

display(reason_comparison)


# ============================================================
# 7. COMPARE SENTIMENT
# ============================================================

print("\n" + "=" * 70)
print("DETECTED_SENTIMENT")
print("=" * 70)

sentiment_comparison = compare_distribution(
    fraud_calls_30,
    non_fraud_calls_30,
    "detected_sentiment"
)

display(sentiment_comparison)


# ============================================================
# 8. COMPARE TRANSCRIPT AVAILABILITY
# ============================================================

print("\n" + "=" * 70)
print("HAS_TRANSCRIPT")
print("=" * 70)

transcript_comparison = compare_distribution(
    fraud_calls_30,
    non_fraud_calls_30,
    "has_transcript"
)

display(transcript_comparison)


# ============================================================
# 9. FIND NON-FRAUD TRANSCRIPTS
# ============================================================

non_fraud_interactions_30 = (
    non_fraud_calls_30[
        [
            "interaction_id",
            "transaction_id",
            "customer_id",
            "days_from_transaction",
        ]
    ]
    .drop_duplicates("interaction_id")
)

non_fraud_transcripts_30 = (
    non_fraud_interactions_30.merge(
        transcripts_df,
        on="interaction_id",
        how="inner"
    )
)

print(
    f"\nNon-fraud transcripts found: "
    f"{len(non_fraud_transcripts_30):,}"
)


# ============================================================
# 10. COMPARE MAIN_TOPICS
# ============================================================

print("\n" + "=" * 70)
print("MAIN_TOPICS")
print("=" * 70)

fraud_topics = (
    fraud_transcripts_30["main_topics"]
    .fillna("NaN")
    .value_counts()
    .rename("fraud_count")
)

non_fraud_topics = (
    non_fraud_transcripts_30["main_topics"]
    .fillna("NaN")
    .value_counts()
    .rename("non_fraud_count")
)

topics_comparison = pd.concat(
    [
        fraud_topics,
        non_fraud_topics
    ],
    axis=1
).fillna(0)

topics_comparison["fraud_pct"] = (
    topics_comparison["fraud_count"]
    / topics_comparison["fraud_count"].sum()
    * 100
)

topics_comparison["non_fraud_pct"] = (
    topics_comparison["non_fraud_count"]
    / topics_comparison["non_fraud_count"].sum()
    * 100
)

topics_comparison["difference_pct_points"] = (
    topics_comparison["fraud_pct"]
    - topics_comparison["non_fraud_pct"]
)

topics_comparison = (
    topics_comparison
    .sort_values(
        "difference_pct_points",
        ascending=False
    )
)

display(topics_comparison)


# ============================================================
# 11. FINAL SUMMARY
# ============================================================

print("\n" + "=" * 70)
print("SUMMARY")
print("=" * 70)

print(
    f"""
Fraud:
- {fraud_transaction_count:,} transactions
- {fraud_with_call:,} with call center ±30 days
- {fraud_with_call / fraud_transaction_count * 100:.2f}% with a call

Non-Fraud:
- {non_fraud_transaction_count:,} transactions
- {non_fraud_with_call:,} with call center ±30 days
- {non_fraud_with_call / non_fraud_transaction_count * 100:.2f}% with a call

Transcripts:
- Fraud: {len(fraud_transcripts_30):,}
- Non-Fraud: {len(non_fraud_transcripts_30):,}
"""
)

FRAUD VS NON-FRAUD CONTACT CENTER COMPARISON
Transactions: 200/1097 archivos
Transactions: 400/1097 archivos
Transactions: 600/1097 archivos
Transactions: 800/1097 archivos
Transactions: 1000/1097 archivos
Transactions: 1097/1097 archivos

Fraudes para comparar: 4,316
No fraudes muestreados: 4,316

GENERAL COMPARISON


,group,transactions,transactions_with_call_30d,pct_with_call_30d,interaction_rows_30d,avg_interactions_per_transaction
0,Fraud,4316,957,22.173309,1092,0.253012
1,Non-Fraud,4316,956,22.150139,1079,0.250000



REASON_CATEGORY


,fraud_count,non_fraud_count,fraud_pct,non_fraud_pct,difference_pct_points
reason_category,,,,,
Comercial,95,64,8.699634,5.931418,2.768216
Técnico,185,163,16.941392,15.106580,1.834812
Retención,29,31,2.655678,2.873031,-0.217353
Queja,180,185,16.483516,17.145505,-0.661989
Transaccional,363,372,33.241758,34.476367,-1.234609
Producto,240,264,21.978022,24.467099,-2.489077



DETECTED_SENTIMENT


,fraud_count,non_fraud_count,fraud_pct,non_fraud_pct,difference_pct_points
detected_sentiment,,,,,
Negativo,182,141,16.666667,13.067655,3.599011
Muy Positivo,41,28,3.754579,2.594995,1.159583
Muy Negativo,57,54,5.219780,5.004634,0.215146
Positivo,107,124,9.798535,11.492122,-1.693588
Neutral,705,732,64.560440,67.840593,-3.280154



HAS_TRANSCRIPT


/tmp/ipykernel_4427/729882798.py:187: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .fillna("NaN")
/tmp/ipykernel_4427/729882798.py:194: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  .fillna("NaN")


,fraud_count,non_fraud_count,fraud_pct,non_fraud_pct,difference_pct_points
has_transcript,,,,,
True,257,252,23.534799,23.354958,0.17984
False,835,827,76.465201,76.645042,-0.17984



Non-fraud transcripts encontrados: 252

MAIN_TOPICS


,fraud_count,non_fraud_count,fraud_pct,non_fraud_pct,difference_pct_points
main_topics,,,,,
Técnico,43,32,16.796875,12.698413,4.098462
Comercial,19,15,7.421875,5.952381,1.469494
Transaccional,97,93,37.890625,36.904762,0.985863
Retención,5,4,1.953125,1.587302,0.365823
Queja,38,43,14.843750,17.063492,-2.219742
Producto,54,65,21.093750,25.793651,-4.699901



RESUMEN

Fraud:
- 4,316 transactions
- 957 con call center ±30 días
- 22.17% con llamada

Non-Fraud:
- 4,316 transactions
- 956 con call center ±30 días
- 22.15% con llamada

Transcripts:
- Fraud: 256
- Non-Fraud: 252

